{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The system\n",
    "=======\n",
    "A system consists mainly of three parts:\n",
    "    - Entities. May represent agents or inactive entities like physical space.\n",
    "    - Links. Represent connections between the entities. \n",
    "    - Time. The passage of time happens.\n",
    "    \n",
    "Basically a system is one big graph that passes information through the links between the nodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "class System:\n",
    "    \"\"\"\n",
    "    The global system and event dispatcher.\n",
    "\n",
    "    Aware of hosts, space, and the passage of time (steps). It is\n",
    "    the main struture of the program, being the only observable and given it\n",
    "    coordinates all processes.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        'Initialize an empty system.'\n",
    "        self.entities = set()\n",
    "        self.entitynames = {}\n",
    "        self.links = {}\n",
    "        self.time = None\n",
    "    # ---\n",
    "\n",
    "    def addEntity(self, entity, name):\n",
    "        'Add an entity to the graph.'\n",
    "        self.entities.add(entity)\n",
    "        self.entitynames[name] = entity\n",
    "    # ---\n",
    "        \n",
    "    def step(self):\n",
    "        'Take a single step forward in time.'\n",
    "        # Process linked items\n",
    "        #TODO...\n",
    "        # Process each entity\n",
    "        for entity in self.entities:\n",
    "            entity.process(self.time)\n",
    "    # ---\n",
    "        \n",
    "        \n",
    "    def run(self, steps):\n",
    "        'Start running the simulation.'\n",
    "        # Init time\n",
    "        if self.time is None:\n",
    "            self.time = 0\n",
    "        # Take steps\n",
    "        for _ in range(steps):\n",
    "            self.step()\n",
    "    # ---\n",
    "    \n",
    "    def stateof(self, entityname):\n",
    "        \"Ask the entity for it's state\"\n",
    "        entity = self.entitynames[entityname]\n",
    "        return entity.logstate()\n",
    "    # ---\n",
    "                \n",
    "                "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Entities\n",
    "=====\n",
    "An entity is something that resides in the system. \n",
    "\n",
    "It processes information according to it's internal state and the information flowing through the links it shares with other entities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Entity:\n",
    "    '''An entity is something that resides in the system.\n",
    "\n",
    "    It processes information according to it's internal state \n",
    "    and the information flowing through the links it shares with \n",
    "    other entities.\n",
    "    \n",
    "    An entity registers the following methods:\n",
    "        - process(time)\n",
    "        - logstate()\n",
    "    '''\n",
    "    \n",
    "    def process(self, time):\n",
    "        pass\n",
    "    \n",
    "    def logstate(self):\n",
    "        pass\n",
    "    \n",
    "# --- Entity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed:  Current cells: 1\n",
      "Processed:  Current cells: 2\n",
      "Processed:  Current cells: 3\n",
      "Processed:  Current cells: 5\n",
      "Processed:  Current cells: 8\n",
      "Processed:  Current cells: 13\n",
      "Processed:  Current cells: 21\n",
      "Processed:  Current cells: 34\n",
      "Processed:  Current cells: 55\n",
      "Processed:  Current cells: 89\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'Current cells: 89'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# < --- Expected use case:\n",
    "\n",
    "class Cells(Entity):\n",
    "    'A special type of group entity that follows a Fibbonacci pattern'\n",
    "    \n",
    "    def __init__(self):\n",
    "        self.current = 1\n",
    "        self.previous = 0\n",
    "        self.changes = {}\n",
    "    \n",
    "    def process(self, time):\n",
    "        c = self.current\n",
    "        p = self.previous\n",
    "        self.current, self.previous = c+p, c\n",
    "        print('Processed: ', self.logstate())\n",
    "        \n",
    "    def logstate(self):\n",
    "        return f'Current cells: {self.current}'\n",
    "# --- Cells\n",
    "\n",
    "\n",
    "# Create a system\n",
    "sys = System()\n",
    "\n",
    "# Add the cells\n",
    "sys.addEntity(Cells(), \n",
    "              name='cells')\n",
    "\n",
    "# Run the processes in the system\n",
    "sys.run(steps=10)\n",
    "\n",
    "# Get the state of the cells\n",
    "sys.stateof('cells')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Interactions\n",
    "=======\n",
    "\n",
    "Multiple entities can interact between them through Interactions.\n",
    "\n",
    "Interactions link 2 or more entities to perform a series of actions together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Interaction:\n",
    "    'A structure representing flow of information btw entities.'\n",
    "    \n",
    "    def __init__(self, entities, effect):\n",
    "        self.entities = entities\n",
    "        self.effects = [effect]\n",
    "        \n",
    "    def append(self, effect):\n",
    "        'Effects btw the same entities can be appended and executed in order'\n",
    "        self.effects.append(effect)\n",
    "# --- Interaction    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to achieve this, the system has to have aditional control for the propagation of the information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class System:\n",
    "    \"\"\"\n",
    "    The global system and event dispatcher.\n",
    "\n",
    "    Aware of hosts, space, and the passage of time (steps). It is\n",
    "    the main struture of the program, being the only observable and given it\n",
    "    coordinates all processes.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        'Initialize an empty system.'\n",
    "        self.entities = set()\n",
    "        self.toentity = {}\n",
    "        self.toentityname = {}\n",
    "        \n",
    "        self.interactions = set()\n",
    "        self.time = None\n",
    "    # ---\n",
    "\n",
    "    def addEntity(self, entity, name):\n",
    "        'Add an entity to the graph.'\n",
    "        self.entities.add(entity)\n",
    "        self.toentity[name] = entity\n",
    "        self.toentityname[entity] = name\n",
    "    # ---\n",
    "        \n",
    "    def step(self):\n",
    "        'Take a single step forward in time.'\n",
    "        # Process linked items\n",
    "        for inter in self.interactions:\n",
    "            entities = inter.entities\n",
    "            for effect in inter.effects:\n",
    "                effect(*entities)\n",
    "        # Process each entity\n",
    "        for entity in self.entities:\n",
    "            entity.process(self.time)\n",
    "    # ---\n",
    "         \n",
    "    def run(self, steps):\n",
    "        'Start running the simulation.'\n",
    "        # Init time\n",
    "        if self.time is None:\n",
    "            self.time = 0\n",
    "        # Take steps\n",
    "        for _ in range(steps):\n",
    "            self.step()\n",
    "    # ---\n",
    "    \n",
    "    def stateof(self, entityname):\n",
    "        \"Ask the entity for it's state\"\n",
    "        entity = self.toentity[entityname]\n",
    "        return entity.logstate()\n",
    "    # ---\n",
    "    \n",
    "    def link(self, effect, entitynames):\n",
    "        'Add a link between two named entities'\n",
    "        entities = tuple(self.toentity[ename] \n",
    "                                for ename in entitynames)\n",
    "        \n",
    "        # Check if there is already some link btw\n",
    "        # those same entities.\n",
    "        if entities in self.interactions:\n",
    "            # Add the new effect\n",
    "            interactions[entities].append(effect)\n",
    "            \n",
    "        else:\n",
    "            # Else, add the new link\n",
    "            interaction = Interaction(entities, effect)\n",
    "            self.interactions.add( interaction )\n",
    "    # ---\n",
    "# --- System\n",
    "                "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below we show a simple use case where interactions play an important role."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total predators:  100\n",
      "Total preys:  100\n",
      "Predators are hunting 50 preys\n",
      "Total preys:  100\n",
      "Total predators:  71\n",
      "Predators are hunting 35 preys\n",
      "Total preys:  128\n",
      "Total predators:  50\n",
      "Predators are hunting 32 preys\n",
      "Total preys:  192\n",
      "Total predators:  35\n",
      "Predators are hunting 33 preys\n",
      "Total preys:  316\n",
      "Total predators:  25\n",
      "Predators are hunting 39 preys\n",
      "Total preys:  552\n",
      "Total predators:  18\n",
      "Predators are hunting 49 preys\n",
      "Total preys:  1004\n",
      "Total predators:  13\n",
      "Predators are hunting 65 preys\n",
      "Total preys:  1876\n",
      "Total predators:  10\n",
      "Predators are hunting 93 preys\n",
      "Total preys:  3564\n",
      "Total predators:  9\n",
      "Predators are hunting 160 preys\n",
      "Total preys:  6806\n",
      "Total predators:  10\n",
      "Predators are hunting 340 preys\n",
      "Total preys:  12930\n",
      "Total predators:  14\n",
      "Predators are hunting 905 preys\n",
      "Total preys:  24048\n",
      "Total predators:  19\n",
      "Predators are hunting 2284 preys\n",
      "Total preys:  43526\n",
      "Total predators:  26\n",
      "Predators are hunting 5658 preys\n",
      "Total preys:  75734\n",
      "Total predators:  36\n",
      "Predators are hunting 13632 preys\n",
      "Total preys:  124202\n",
      "Total predators:  50\n",
      "Predators are hunting 31050 preys\n",
      "Total preys:  186302\n",
      "Total predators:  70\n",
      "Predators are hunting 65205 preys\n",
      "Total preys:  242192\n",
      "Total predators:  98\n",
      "Predators are hunting 118674 preys\n",
      "Total preys:  247034\n",
      "Total predators:  137\n",
      "Predators are hunting 169218 preys\n",
      "Total preys:  155630\n",
      "Total predators:  191\n",
      "Predators are hunting 148626 preys\n",
      "Total preys:  14006\n",
      "Total predators:  267\n",
      "Predators are hunting 18698 preys\n",
      "Total preys:  0\n",
      "Total predators:  373\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  261\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  182\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  127\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  88\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  61\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  42\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  29\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  20\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  14\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  9\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  6\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  4\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  2\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  1\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n",
      "Predators are hunting 0 preys\n",
      "Total preys:  0\n",
      "Total predators:  0\n"
     ]
    }
   ],
   "source": [
    "# < --- Intended use case:\n",
    "\n",
    "class Animals(Entity):\n",
    "    'A simple group entity.'\n",
    "    \n",
    "    def __init__(self, pop=10):\n",
    "        self.hunt_results = None\n",
    "        self.population = pop\n",
    "    # ---\n",
    "        \n",
    "    @property\n",
    "    def population(self):\n",
    "        return self._population\n",
    "    # ---\n",
    "    \n",
    "    @population.setter\n",
    "    def population(self, value):\n",
    "        'The population consists of integers >= 0'\n",
    "        if value < 0:\n",
    "            self._population = 0\n",
    "        else:\n",
    "            self._population = int(value)\n",
    "    # ---\n",
    "                                   \n",
    "    def logstate(self):\n",
    "        return self.population\n",
    "    # ---\n",
    "# ---\n",
    "\n",
    "def limit(quantity, interval):\n",
    "    bottom, top = interval\n",
    "    if quantity > top:\n",
    "        quantity = top\n",
    "        \n",
    "    elif quantity < bottom:\n",
    "        quantity = bottom\n",
    "        \n",
    "    return quantity\n",
    "# ---\n",
    "        \n",
    "    \n",
    "class Predators(Animals):\n",
    "    def process(self, time):\n",
    "        self.population += self.hunt_results\n",
    "        self.hunt_results = None\n",
    "        self.die()\n",
    "        self.logstate()\n",
    "        \n",
    "    def die(self):\n",
    "        self.population -= 0.3*self.population\n",
    "        \n",
    "    def logstate(self):\n",
    "        print('Total predators: ', self.population)\n",
    "# ---\n",
    "\n",
    "class Preys(Animals):\n",
    "    def process(self, time):\n",
    "        # Reduce population due to hunting\n",
    "        self.population -= self.hunt_results\n",
    "        self.hunt_results = None\n",
    "        self.reproduce()\n",
    "        self.logstate()\n",
    "        \n",
    "    def logstate(self):\n",
    "        print('Total preys: ', self.population)\n",
    "        \n",
    "    def reproduce(self):\n",
    "        self.population *= 2\n",
    "# ---\n",
    "            \n",
    "\n",
    "def predation(predators, preys):\n",
    "    '''The predation interaction.\n",
    "    \n",
    "    h (hunting) is the effect of predators in prey.\n",
    "    f (feeding) is the effect of prey in predators.\n",
    "    '''\n",
    "    h, f = 0.005, 0.0002\n",
    "    close_encounters = preys.population * predators.population\n",
    "    \n",
    "    preys.hunt_results = h * close_encounters\n",
    "    predators.hunt_results = limit(f * close_encounters,\n",
    "                                   [0, predators.population])\n",
    "    \n",
    "    print(f'Predators are hunting {int(preys.hunt_results)} preys')\n",
    "# ---\n",
    "    \n",
    "\n",
    "# Create the system\n",
    "sys = System()\n",
    "\n",
    "# Create the cells\n",
    "sys.addEntity(Predators(pop=100),\n",
    "              name='Predators')\n",
    "sys.addEntity(Preys(pop=100),\n",
    "              name='Preys')\n",
    "\n",
    "# Add the link btw the cells\n",
    "sys.link(predation, \n",
    "         entitynames=['Predators', 'Preys'])\n",
    "\n",
    "# Verify the initial state\n",
    "sys.stateof('Predators')\n",
    "sys.stateof('Preys')\n",
    "\n",
    "# Run the simulation\n",
    "sys.run(steps=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The use case we are mainly interested in is in simulating several cell types along with cellular automata fields for chemicals, so we proceed to move forwards towards that intended use case"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Simple Cellular Automaton\n",
    "===============\n",
    "\n",
    "Now we tackle the problem of a simple 1D CA within our framework."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "........................................#.......................................\n",
      ".......................................#.#......................................\n",
      "......................................#...#.....................................\n",
      ".....................................#.#.#.#....................................\n",
      "....................................#.......#...................................\n",
      "...................................#.#.....#.#..................................\n",
      "..................................#...#...#...#.................................\n",
      ".................................#.#.#.#.#.#.#.#................................\n",
      "................................#...............#...............................\n",
      "...............................#.#.............#.#..............................\n",
      "..............................#...#...........#...#.............................\n",
      ".............................#.#.#.#.........#.#.#.#............................\n",
      "............................#.......#.......#.......#...........................\n",
      "...........................#.#.....#.#.....#.#.....#.#..........................\n",
      "..........................#...#...#...#...#...#...#...#.........................\n",
      ".........................#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#........................\n",
      "........................#...............................#.......................\n",
      ".......................#.#.............................#.#......................\n",
      "......................#...#...........................#...#.....................\n",
      ".....................#.#.#.#.........................#.#.#.#....................\n",
      "....................#.......#.......................#.......#...................\n",
      "...................#.#.....#.#.....................#.#.....#.#..................\n",
      "..................#...#...#...#...................#...#...#...#.................\n",
      ".................#.#.#.#.#.#.#.#.................#.#.#.#.#.#.#.#................\n",
      "................#...............#...............#...............#...............\n",
      "...............#.#.............#.#.............#.#.............#.#..............\n",
      "..............#...#...........#...#...........#...#...........#...#.............\n",
      ".............#.#.#.#.........#.#.#.#.........#.#.#.#.........#.#.#.#............\n",
      "............#.......#.......#.......#.......#.......#.......#.......#...........\n",
      "...........#.#.....#.#.....#.#.....#.#.....#.#.....#.#.....#.#.....#.#..........\n",
      "..........#...#...#...#...#...#...#...#...#...#...#...#...#...#...#...#.........\n",
      ".........#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#........\n",
      "........#...............................................................#.......\n",
      ".......#.#.............................................................#.#......\n",
      "......#...#...........................................................#...#.....\n",
      ".....#.#.#.#.........................................................#.#.#.#....\n",
      "....#.......#.......................................................#.......#...\n",
      "...#.#.....#.#.....................................................#.#.....#.#..\n",
      "..#...#...#...#...................................................#...#...#...#.\n",
      ".#.#.#.#.#.#.#.#.................................................#.#.#.#.#.#.#..\n",
      "................#...............................................#.............#.\n",
      "...............#.#.............................................#.#...........#..\n",
      "..............#...#...........................................#...#.........#.#.\n",
      ".............#.#.#.#.........................................#.#.#.#.......#....\n",
      "............#.......#.......................................#.......#.....#.#...\n",
      "...........#.#.....#.#.....................................#.#.....#.#...#...#..\n",
      "..........#...#...#...#...................................#...#...#...#.#.#.#.#.\n",
      ".........#.#.#.#.#.#.#.#.................................#.#.#.#.#.#.#..........\n",
      "........#...............#...............................#.............#.........\n",
      ".......#.#.............#.#.............................#.#...........#.#........\n",
      "......#...#...........#...#...........................#...#.........#...#.......\n"
     ]
    }
   ],
   "source": [
    "class CellularAutomaton(Entity):\n",
    "    'A simple cellular automaton.'\n",
    "    \n",
    "    def __init__(self, n=25):\n",
    "        # Place a single active cell in the middle\n",
    "        self.gridstate = tuple(1 if i == n//2 else 0\n",
    "                                   for i in range(n))\n",
    "        self.rule = [0, 1, 0, 1, 1, 0, 1, 0]\n",
    "        \n",
    "    def process(self, time):\n",
    "        self.gridstate = self.applyrule(self.gridstate, self.rule)\n",
    "        print(''.join(self.logstate()))\n",
    "        \n",
    "    def applyrule(self, grid, rule):\n",
    "        newgrid = list(grid)\n",
    "        for i,cell in enumerate(grid):\n",
    "            # Cells at the edges are constant\n",
    "            if 1 <= i < len(grid)-1:\n",
    "                neighbors = grid[i-1:i+2]\n",
    "                newgrid[i] = self.rule[self.neighbor_n(neighbors)]\n",
    "        return tuple(newgrid)\n",
    "                \n",
    "    def neighbor_n(self, neighbors):\n",
    "        a, b, c = neighbors\n",
    "        return c + b*2 + a*4\n",
    "    \n",
    "    def logstate(self):\n",
    "        return [ '#' if cell else '.' for cell in self.gridstate ]\n",
    "    \n",
    "sys = System()\n",
    "\n",
    "sys.addEntity(CellularAutomaton(80),\n",
    "              name='Sierpinsky')\n",
    "\n",
    "print( ''.join(sys.stateof('Sierpinsky')) )\n",
    "\n",
    "sys.run(steps=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We note now (with the contortions in the developer side on `CellularAutomaton.process`, line 12 and on the user side in the line 35, only to print the system state), that it is important to be available to get a representation (as __repr__ and __str__) of the system itself, as for the entities and interactions, as well as to expose the real state of the entities. So, we make further changes to the system."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Entity:\n",
    "    '''An entity is something that resides in the system.\n",
    "\n",
    "    It processes information according to it's internal state \n",
    "    and the information flowing through the links it shares with \n",
    "    other entities.\n",
    "    \n",
    "    An entity registers the following methods:\n",
    "        - process(time)\n",
    "    '''\n",
    "    def __init__(self):\n",
    "        self.state = None\n",
    "    \n",
    "    def process(self, time):\n",
    "        pass\n",
    "# --- Entity\n",
    "\n",
    "\n",
    "\n",
    "class System:\n",
    "    \"\"\"\n",
    "    The global system and event dispatcher.\n",
    "\n",
    "    Aware of hosts, space, and the passage of time (steps). It is\n",
    "    the main struture of the program, being the only observable and given it\n",
    "    coordinates all processes.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        'Initialize an empty system.'\n",
    "        self.entities = set()\n",
    "        self.toentity = {}\n",
    "        self.toentityname = {}\n",
    "        \n",
    "        self.interactions = set()\n",
    "        self.time = None\n",
    "    # ---\n",
    "    \n",
    "    def __getitem__(self, entityname):\n",
    "        'Access the entities by name.'\n",
    "        return self.toentity[entityname]\n",
    "    # ---\n",
    "\n",
    "    def add(self, entity, name):\n",
    "        'Add an entity to the graph.'\n",
    "        self.entities.add(entity)\n",
    "        self.toentity[name] = entity\n",
    "        self.toentityname[entity] = name\n",
    "    # ---\n",
    "        \n",
    "    def step(self):\n",
    "        'Take a single step forward in time.'\n",
    "        # Process linked items\n",
    "        for inter in self.interactions:\n",
    "            entities = inter.entities\n",
    "            for effect in inter.effects:\n",
    "                effect(*entities)\n",
    "        # Process each entity\n",
    "        for entity in self.entities:\n",
    "            entity.process(self.time)\n",
    "    # ---\n",
    "    \n",
    "        \n",
    "    def start(self):\n",
    "        'Initialize things to be initialized.'\n",
    "        # Init time\n",
    "        if self.time is None:\n",
    "            self.time = 0\n",
    "    # ---\n",
    "         \n",
    "    def run(self, steps):\n",
    "        'Start running the simulation.'\n",
    "        # Make sure things are initialized \n",
    "        self.start()\n",
    "        \n",
    "        # Take steps\n",
    "        for _ in range(steps):\n",
    "            self.step()\n",
    "    # ---\n",
    "    \n",
    "    def stateof(self, entityname):\n",
    "        \"Ask the entity for it's state\"\n",
    "        return self[entityname].state\n",
    "    # ---\n",
    "    \n",
    "    def link(self, effect, entitynames):\n",
    "        'Add a link between named entities.'\n",
    "        entities = tuple(self.toentity[ename] \n",
    "                                for ename in entitynames)\n",
    "        \n",
    "        # Check if there is already some link btw\n",
    "        # those same entities.\n",
    "        if entities in self.interactions:\n",
    "            # Add the new effect\n",
    "            interactions[entities].append(effect)\n",
    "            \n",
    "        else:\n",
    "            # Else, add the new link\n",
    "            interaction = Interaction(entities, effect)\n",
    "            self.interactions.add( interaction )\n",
    "    # ---\n",
    "# --- System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "........................................#.......................................\n",
      ".......................................#.#......................................\n",
      "......................................#...#.....................................\n",
      ".....................................#.#.#.#....................................\n",
      "....................................#.......#...................................\n",
      "...................................#.#.....#.#..................................\n",
      "..................................#...#...#...#.................................\n",
      ".................................#.#.#.#.#.#.#.#................................\n",
      "................................#...............#...............................\n",
      "...............................#.#.............#.#..............................\n",
      "..............................#...#...........#...#.............................\n",
      ".............................#.#.#.#.........#.#.#.#............................\n",
      "............................#.......#.......#.......#...........................\n",
      "...........................#.#.....#.#.....#.#.....#.#..........................\n",
      "..........................#...#...#...#...#...#...#...#.........................\n",
      ".........................#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#........................\n",
      "........................#...............................#.......................\n",
      ".......................#.#.............................#.#......................\n",
      "......................#...#...........................#...#.....................\n",
      ".....................#.#.#.#.........................#.#.#.#....................\n",
      "....................#.......#.......................#.......#...................\n",
      "...................#.#.....#.#.....................#.#.....#.#..................\n",
      "..................#...#...#...#...................#...#...#...#.................\n",
      ".................#.#.#.#.#.#.#.#.................#.#.#.#.#.#.#.#................\n",
      "................#...............#...............#...............#...............\n",
      "...............#.#.............#.#.............#.#.............#.#..............\n",
      "..............#...#...........#...#...........#...#...........#...#.............\n",
      ".............#.#.#.#.........#.#.#.#.........#.#.#.#.........#.#.#.#............\n",
      "............#.......#.......#.......#.......#.......#.......#.......#...........\n",
      "...........#.#.....#.#.....#.#.....#.#.....#.#.....#.#.....#.#.....#.#..........\n",
      "..........#...#...#...#...#...#...#...#...#...#...#...#...#...#...#...#.........\n",
      ".........#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#........\n",
      "........#...............................................................#.......\n",
      ".......#.#.............................................................#.#......\n",
      "......#...#...........................................................#...#.....\n",
      ".....#.#.#.#.........................................................#.#.#.#....\n",
      "....#.......#.......................................................#.......#...\n",
      "...#.#.....#.#.....................................................#.#.....#.#..\n",
      "..#...#...#...#...................................................#...#...#...#.\n",
      ".#.#.#.#.#.#.#.#.................................................#.#.#.#.#.#.#..\n",
      "................#...............................................#.............#.\n",
      "...............#.#.............................................#.#...........#..\n",
      "..............#...#...........................................#...#.........#.#.\n",
      ".............#.#.#.#.........................................#.#.#.#.......#....\n",
      "............#.......#.......................................#.......#.....#.#...\n",
      "...........#.#.....#.#.....................................#.#.....#.#...#...#..\n",
      "..........#...#...#...#...................................#...#...#...#.#.#.#.#.\n",
      ".........#.#.#.#.#.#.#.#.................................#.#.#.#.#.#.#..........\n",
      "........#...............#...............................#.............#.........\n",
      ".......#.#.............#.#.............................#.#...........#.#........\n",
      "......#...#...........#...#...........................#...#.........#...#.......\n"
     ]
    }
   ],
   "source": [
    "# Intended use case...\n",
    "class CellularAutomaton(Entity):\n",
    "    'A simple cellular automaton.'\n",
    "    \n",
    "    def __init__(self, n=25):\n",
    "        # Place a single active cell in the middle\n",
    "        self.gridstate = tuple(1 if i == n//2 else 0\n",
    "                                   for i in range(n))\n",
    "        self.rule = [0, 1, 0, 1, 1, 0, 1, 0]\n",
    "    # ---\n",
    "        \n",
    "    @property\n",
    "    def state(self):\n",
    "        return [ '#' if cell else '.' for cell in self.gridstate ]\n",
    "    # ---\n",
    "        \n",
    "    def process(self, time):\n",
    "        'Move one time step forward.'\n",
    "        self.gridstate = self.applyrule(self.gridstate)\n",
    "        print(''.join(self.state))\n",
    "    # ---\n",
    "        \n",
    "    def applyrule(self, grid):\n",
    "        'Apply automaton rule to the given grid.'\n",
    "        newgrid = list(grid)\n",
    "        for i,cell in enumerate(grid):\n",
    "            # Cells at the edges are constant\n",
    "            if 1 <= i < len(grid)-1:\n",
    "                neighbors = grid[i-1:i+2]\n",
    "                newgrid[i] = self.rule[self.process_neighbor(neighbors)]\n",
    "        # Return an immutable new grid\n",
    "        return tuple(newgrid)\n",
    "                \n",
    "    def process_neighbor(self, neighbors):\n",
    "        'Convert neighborhood to integer to apply rule.'\n",
    "        a, b, c = neighbors\n",
    "        return c + b*2 + a*4\n",
    "    \n",
    "    def __str__(self):\n",
    "        return ''.join(self.state)\n",
    "# --- CellularAutomaton\n",
    "    \n",
    "\n",
    "# < --- ---    \n",
    "\n",
    "\n",
    "sys = System()\n",
    "\n",
    "sys.add(CellularAutomaton(80),\n",
    "              name='Sierpinsky')\n",
    "\n",
    "print(sys['Sierpinsky'] )\n",
    "\n",
    "sys.run(steps=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We made an important change to the Entity object interface, as it can be seen, the method `logstate` was eliminated, and now the Entity state can and should be accessed with the `state` attribute. From this, now \n",
    "```python\n",
    "    sys.stateof('SomeEntity') == sys['SomeEntity'].state == sys.toentity['SomeEntity'].state\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, it is desirable to have the ability to modify the run behavior on the fly, so, one option is to allow for hooks in the run function and on adding the automaton."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import collections\n",
    "\n",
    "class Interaction:\n",
    "    'A structure representing flow of information btw entities.'\n",
    "    \n",
    "    def __init__(self, entities, effect):\n",
    "        self.entities = entities\n",
    "        self.effects = [effect]\n",
    "        \n",
    "    def append(self, effect):\n",
    "        'Effects btw the same entities can be appended and executed in order'\n",
    "        self.effects.append(effect)\n",
    "        \n",
    "    def process(self):\n",
    "        'Executes the interaction.'\n",
    "        for effect in self.effects:\n",
    "            effect(*self.entities)\n",
    "# --- Interaction \n",
    "\n",
    "\n",
    "\n",
    "class System:\n",
    "    \"\"\"\n",
    "    The global system and event dispatcher.\n",
    "\n",
    "    Aware of hosts, space, and the passage of time (steps). It is\n",
    "    the main struture of the program, being the only observable and given it\n",
    "    coordinates all processes.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        'Initialize an empty system.'\n",
    "        self.entities = set()\n",
    "        self.toentity = {}\n",
    "        self.toentityname = {}\n",
    "        \n",
    "        self.interactions = {}\n",
    "        self.time = None\n",
    "        \n",
    "        self.inithooks = {}\n",
    "        self.prehooks = {}\n",
    "        self.hooks = {}\n",
    "    # ---\n",
    "    \n",
    "    def __getitem__(self, entityname):\n",
    "        'Access the entities by name.'\n",
    "        return self.toentity[entityname]\n",
    "    # ---\n",
    "\n",
    "    def add(self, entity, name, inithook=None):\n",
    "        'Add an entity to the graph.'\n",
    "        self.entities.add(entity)\n",
    "        self.toentity[name] = entity\n",
    "        self.toentityname[entity] = name\n",
    "        # Process hooks\n",
    "        if inithook:\n",
    "            self.add_interaction_to(self.inithooks,\n",
    "                                    inithook, \n",
    "                                    [name])\n",
    "    # ---\n",
    "    \n",
    "    def process_interactions_in(self, interactions):\n",
    "        'From the dict-like container of interactions, process items.'\n",
    "        for interaction in interactions.values():\n",
    "            interaction.process()\n",
    "    # ---\n",
    "        \n",
    "    def step(self):\n",
    "        'Take a single step forward in time.'\n",
    "        # Process pre-step hooks\n",
    "        self.process_interactions_in(self.prehooks)\n",
    "        \n",
    "        # Process linked items\n",
    "        self.process_interactions_in(self.interactions)\n",
    "        # Process each entity\n",
    "        for entity in self.entities:\n",
    "            entity.process(self.time)\n",
    "            \n",
    "        # Process post-step hooks\n",
    "        self.process_interactions_in(self.hooks)\n",
    "    # ---\n",
    "         \n",
    "    def start(self):\n",
    "        'Initialize things before starting simulation.'\n",
    "        # Init time\n",
    "        if self.time is None:\n",
    "            self.time = 0\n",
    "        # Make the initialization actions\n",
    "        self.process_interactions_in(self.inithooks)\n",
    "    # ---\n",
    "    \n",
    "    def update_hooks(self, hooktype, newhooks):\n",
    "        'Add the given hooks to the system.'\n",
    "        if newhooks:\n",
    "            for hook, entitynames in newhooks:\n",
    "                self.add_interaction_to(hooktype, hook, entitynames)\n",
    "    # ---\n",
    "         \n",
    "    def run(self, steps, inithooks=None, hooks=None, prehooks=None):\n",
    "        'Start running the simulation.'\n",
    "        # Handle hooks\n",
    "        self.update_hooks(self.inithooks, inithooks)\n",
    "        self.update_hooks(self.hooks, hooks)\n",
    "        self.update_hooks(self.prehooks, prehooks)\n",
    "        \n",
    "        # Make sure things are initialized \n",
    "        self.start()\n",
    "        \n",
    "        # Take steps\n",
    "        for _ in range(steps):\n",
    "            self.step() \n",
    "    # ---\n",
    "    \n",
    "    def stateof(self, entityname):\n",
    "        \"Ask the entity for it's state\"\n",
    "        return self[entityname].state\n",
    "    # ---\n",
    "    \n",
    "    def link(self, effect, entitynames):\n",
    "        'Add an interation btw named entities.'\n",
    "        self.add_interaction_to(self.interactions,\n",
    "                                effect, \n",
    "                                entitynames)\n",
    "    # ---\n",
    "    \n",
    "    def add_interaction_to(self, container, effect, entitynames):\n",
    "        'Add an interaction btw named entities to a dict-like container.'\n",
    "        entities = tuple(self.toentity[ename] \n",
    "                                for ename in entitynames)\n",
    "        \n",
    "        # Check if there is already some link btw\n",
    "        # those same entities.\n",
    "        if entities in container:\n",
    "            # Add the new effect\n",
    "            container[entities].append(effect)\n",
    "            \n",
    "        else:\n",
    "            # Else, add the new link\n",
    "            interaction = Interaction(entities, effect)\n",
    "            container[entities] = interaction\n",
    "    # ---\n",
    "# --- System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "..................................................#.................................................\n",
      ".................................................#.#................................................\n",
      "................................................#...#...............................................\n",
      "...............................................#.#.#.#..............................................\n",
      "..............................................#.......#.............................................\n",
      ".............................................#.#.....#.#............................................\n",
      "............................................#...#...#...#...........................................\n",
      "...........................................#.#.#.#.#.#.#.#..........................................\n",
      "..........................................#...............#.........................................\n",
      ".........................................#.#.............#.#........................................\n",
      "........................................#...#...........#...#.......................................\n",
      ".......................................#.#.#.#.........#.#.#.#......................................\n",
      "......................................#.......#.......#.......#.....................................\n",
      ".....................................#.#.....#.#.....#.#.....#.#....................................\n",
      "....................................#...#...#...#...#...#...#...#...................................\n",
      "...................................#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#..................................\n",
      "..................................#...............................#.................................\n",
      ".................................#.#.............................#.#................................\n",
      "................................#...#...........................#...#...............................\n",
      "...............................#.#.#.#.........................#.#.#.#..............................\n",
      "..............................#.......#.......................#.......#.............................\n",
      ".............................#.#.....#.#.....................#.#.....#.#............................\n",
      "............................#...#...#...#...................#...#...#...#...........................\n",
      "...........................#.#.#.#.#.#.#.#.................#.#.#.#.#.#.#.#..........................\n",
      "..........................#...............#...............#...............#.........................\n",
      ".........................#.#.............#.#.............#.#.............#.#........................\n",
      "........................#...#...........#...#...........#...#...........#...#.......................\n",
      ".......................#.#.#.#.........#.#.#.#.........#.#.#.#.........#.#.#.#......................\n",
      "......................#.......#.......#.......#.......#.......#.......#.......#.....................\n",
      ".....................#.#.....#.#.....#.#.....#.#.....#.#.....#.#.....#.#.....#.#....................\n",
      "....................#...#...#...#...#...#...#...#...#...#...#...#...#...#...#...#...................\n",
      "...................#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#..................\n",
      "..................#...............................................................#.................\n",
      ".................#.#.............................................................#.#................\n",
      "................#...#...........................................................#...#...............\n",
      "...............#.#.#.#.........................................................#.#.#.#..............\n",
      "..............#.......#.......................................................#.......#.............\n",
      ".............#.#.....#.#.....................................................#.#.....#.#............\n",
      "............#...#...#...#...................................................#...#...#...#...........\n",
      "...........#.#.#.#.#.#.#.#.................................................#.#.#.#.#.#.#.#..........\n",
      "..........#...............#...............................................#...............#.........\n",
      ".........#.#.............#.#.............................................#.#.............#.#........\n",
      "........#...#...........#...#...........................................#...#...........#...#.......\n",
      ".......#.#.#.#.........#.#.#.#.........................................#.#.#.#.........#.#.#.#......\n",
      "......#.......#.......#.......#.......................................#.......#.......#.......#.....\n",
      ".....#.#.....#.#.....#.#.....#.#.....................................#.#.....#.#.....#.#.....#.#....\n",
      "....#...#...#...#...#...#...#...#...................................#...#...#...#...#...#...#...#...\n",
      "...#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.................................#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#..\n",
      "..#...............................#...............................#...............................#.\n",
      ".#.#.............................#.#.............................#.#.............................#..\n",
      "Final state:  ....#...........................#...#...........................#...#...........................#.#.\n",
      "['.', '.', '.', '.', '#', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '#', '.', '.', '.', '#', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '#', '.', '.', '.', '#', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '#', '.', '#', '.']\n"
     ]
    }
   ],
   "source": [
    "# Intended use case...\n",
    "class SilentCA(CellularAutomaton):\n",
    "    'A cellular automaton that does not print to the screen.'\n",
    "        \n",
    "    def process(self, time):\n",
    "        'Move one time step forward, do not print state.'\n",
    "        self.gridstate = self.applyrule(self.gridstate)\n",
    "    # ---\n",
    "# --- CellularAutomaton\n",
    "\n",
    "\n",
    "def simplelogging(entity):\n",
    "    'We will use this to print the state.'\n",
    "    print(entity)\n",
    "# ---\n",
    "\n",
    "\n",
    "## < --- --- ----\n",
    "\n",
    "\n",
    "sys = System()\n",
    "\n",
    "sys.add(SilentCA(100), \n",
    "        name='Sierpinsky')\n",
    "\n",
    "sys.run(steps=50, \n",
    "        prehooks=[(simplelogging,['Sierpinsky'])] )\n",
    "\n",
    "print('Final state: ', sys['Sierpinsky'])\n",
    "print(sys.stateof('Sierpinsky'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<__main__.System at 0x7f139c568668>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{<__main__.SilentCA at 0x7f139c5689e8>}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sys.entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on SilentCA in module __main__ object:\n",
      "\n",
      "class SilentCA(CellularAutomaton)\n",
      " |  A cellular automaton that does not print to the screen.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      SilentCA\n",
      " |      CellularAutomaton\n",
      " |      Entity\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  process(self, time)\n",
      " |      Move one time step forward, do not print state.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from CellularAutomaton:\n",
      " |  \n",
      " |  __init__(self, n=25)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return str(self).\n",
      " |  \n",
      " |  applyrule(self, grid)\n",
      " |      Apply automaton rule to the given grid.\n",
      " |  \n",
      " |  process_neighbor(self, neighbors)\n",
      " |      Convert neighborhood to integer to apply rule.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from CellularAutomaton:\n",
      " |  \n",
      " |  state\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Entity:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for e in sys.entities:\n",
    "    help(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see now that several changes have been done to the code. The hooks where added with three flavors: inithooks (executed before the simulation starts), prehooks (executed before any step) and hooks (executed at the end of every step). This developed in a major factorization of how the interactions are applied, as the hooks are implemented using the same mechanism."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's clean up the API, so that the usage becomes more intuitive."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import collections\n",
    "\n",
    "class Interaction:\n",
    "    'A structure representing flow of information btw entities.'\n",
    "    \n",
    "    def __init__(self, entities, effect):\n",
    "        self.entities = entities\n",
    "        self.effects = [effect]\n",
    "        \n",
    "    def append(self, effect):\n",
    "        'Effects btw the same entities can be appended and executed in order'\n",
    "        self.effects.append(effect)\n",
    "        \n",
    "    def process(self):\n",
    "        'Executes the interaction.'\n",
    "        for effect in self.effects:\n",
    "            effect(*self.entities)\n",
    "# --- Interaction\n",
    "\n",
    "'Representing an isolated process.'\n",
    "Process = collections.namedtuple('Process', ['entities', 'effects'])\n",
    "# --- Process\n",
    "\n",
    "\n",
    "class System:\n",
    "    \"\"\"\n",
    "    The global system and event dispatcher.\n",
    "\n",
    "    Aware of hosts, space, and the passage of time (steps). It is\n",
    "    the main struture of the program, being the only observable and given it\n",
    "    coordinates all processes.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        'Initialize an empty system.'\n",
    "        self.entities = set()\n",
    "        self.toentity = {}\n",
    "        self.toentityname = {}\n",
    "        \n",
    "        self.interactions = {}\n",
    "        self.time = None\n",
    "        \n",
    "        self.inithooks = {}\n",
    "        self.prehooks = {}\n",
    "        self.hooks = {}\n",
    "    # ---\n",
    "    \n",
    "    def __getitem__(self, entityname):\n",
    "        'Access the entities by name.'\n",
    "        return self.toentity[entityname]\n",
    "    # ---\n",
    "\n",
    "    def add(self, entity, name, inithook=None):\n",
    "        'Add an entity to the graph.'\n",
    "        self.entities.add(entity)\n",
    "        self.toentity[name] = entity\n",
    "        self.toentityname[entity] = name\n",
    "        # Process hooks\n",
    "        if inithook:\n",
    "            self.add_interaction_to(self.inithooks,\n",
    "                                    inithook, \n",
    "                                    [name])\n",
    "    # ---\n",
    "    \n",
    "    def process_interactions_in(self, interactions):\n",
    "        'From the dict-like container of interactions, process items.'\n",
    "        for interaction in interactions.values():\n",
    "            interaction.process()\n",
    "    # ---\n",
    "        \n",
    "    def step(self):\n",
    "        'Take a single step forward in time.'\n",
    "        # Process pre-step hooks\n",
    "        self.process_interactions_in(self.prehooks)\n",
    "        \n",
    "        # Process linked items\n",
    "        self.process_interactions_in(self.interactions)\n",
    "        # Process each entity\n",
    "        for entity in self.entities:\n",
    "            entity.process(self.time)\n",
    "            \n",
    "        # Process post-step hooks\n",
    "        self.process_interactions_in(self.hooks)\n",
    "    # ---\n",
    "         \n",
    "    def start(self):\n",
    "        'Initialize things before starting simulation.'\n",
    "        # Init time\n",
    "        if self.time is None:\n",
    "            self.time = 0\n",
    "        # Make the initialization actions\n",
    "        self.process_interactions_in(self.inithooks)\n",
    "    # ---\n",
    "    \n",
    "    def update_hooks(self, hooktype, newhooks):\n",
    "        'Add the given hooks to the system.'\n",
    "        if newhooks:\n",
    "            for hook, entitynames in newhooks:\n",
    "                self.add_interaction_to(hooktype, hook, entitynames)\n",
    "    # ---\n",
    "         \n",
    "    def run(self, steps, init=None, after_step=None, before_step=None):\n",
    "        'Start running the simulation.'\n",
    "        # Handle hooks\n",
    "        self.update_hooks(self.inithooks, init)\n",
    "        self.update_hooks(self.hooks, after_step)\n",
    "        self.update_hooks(self.prehooks, before_step)\n",
    "        \n",
    "        # Make sure things are initialized \n",
    "        self.start()\n",
    "        \n",
    "        # Take steps\n",
    "        for _ in range(steps):\n",
    "            self.step() \n",
    "    # ---\n",
    "    \n",
    "    def stateof(self, entityname):\n",
    "        \"Ask the entity for it's state\"\n",
    "        return self[entityname].state\n",
    "    # ---\n",
    "    \n",
    "    def link(self, effect, entitynames):\n",
    "        'Add an interation btw named entities.'\n",
    "        self.add_interaction_to(self.interactions,\n",
    "                                effect, \n",
    "                                entitynames)\n",
    "    # ---\n",
    "    \n",
    "    def add_interaction_to(self, container, effect, entitynames):\n",
    "        'Add an interaction btw named entities to a dict-like container.'\n",
    "        entities = tuple(self.toentity[ename] \n",
    "                                for ename in entitynames)\n",
    "        \n",
    "        # Check if there is already some link btw\n",
    "        # those same entities.\n",
    "        if entities in container:\n",
    "            # Add the new effect\n",
    "            container[entities].append(effect)\n",
    "            \n",
    "        else:\n",
    "            # Else, add the new link\n",
    "            interaction = Interaction(entities, effect)\n",
    "            container[entities] = interaction\n",
    "    # ---\n",
    "# --- System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "..................................................#.................................................\n",
      ".................................................#.#................................................\n",
      "................................................#...#...............................................\n",
      "...............................................#.#.#.#..............................................\n",
      "..............................................#.......#.............................................\n",
      ".............................................#.#.....#.#............................................\n",
      "............................................#...#...#...#...........................................\n",
      "...........................................#.#.#.#.#.#.#.#..........................................\n",
      "..........................................#...............#.........................................\n",
      ".........................................#.#.............#.#........................................\n",
      "........................................#...#...........#...#.......................................\n",
      ".......................................#.#.#.#.........#.#.#.#......................................\n",
      "......................................#.......#.......#.......#.....................................\n",
      ".....................................#.#.....#.#.....#.#.....#.#....................................\n",
      "....................................#...#...#...#...#...#...#...#...................................\n",
      "...................................#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#..................................\n",
      "..................................#...............................#.................................\n",
      ".................................#.#.............................#.#................................\n",
      "................................#...#...........................#...#...............................\n",
      "...............................#.#.#.#.........................#.#.#.#..............................\n",
      "..............................#.......#.......................#.......#.............................\n",
      ".............................#.#.....#.#.....................#.#.....#.#............................\n",
      "............................#...#...#...#...................#...#...#...#...........................\n",
      "...........................#.#.#.#.#.#.#.#.................#.#.#.#.#.#.#.#..........................\n",
      "..........................#...............#...............#...............#.........................\n",
      ".........................#.#.............#.#.............#.#.............#.#........................\n",
      "........................#...#...........#...#...........#...#...........#...#.......................\n",
      ".......................#.#.#.#.........#.#.#.#.........#.#.#.#.........#.#.#.#......................\n",
      "......................#.......#.......#.......#.......#.......#.......#.......#.....................\n",
      ".....................#.#.....#.#.....#.#.....#.#.....#.#.....#.#.....#.#.....#.#....................\n",
      "....................#...#...#...#...#...#...#...#...#...#...#...#...#...#...#...#...................\n",
      "...................#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#..................\n",
      "..................#...............................................................#.................\n",
      ".................#.#.............................................................#.#................\n",
      "................#...#...........................................................#...#...............\n",
      "...............#.#.#.#.........................................................#.#.#.#..............\n",
      "..............#.......#.......................................................#.......#.............\n",
      ".............#.#.....#.#.....................................................#.#.....#.#............\n",
      "............#...#...#...#...................................................#...#...#...#...........\n",
      "...........#.#.#.#.#.#.#.#.................................................#.#.#.#.#.#.#.#..........\n",
      "..........#...............#...............................................#...............#.........\n",
      ".........#.#.............#.#.............................................#.#.............#.#........\n",
      "........#...#...........#...#...........................................#...#...........#...#.......\n",
      ".......#.#.#.#.........#.#.#.#.........................................#.#.#.#.........#.#.#.#......\n",
      "......#.......#.......#.......#.......................................#.......#.......#.......#.....\n",
      ".....#.#.....#.#.....#.#.....#.#.....................................#.#.....#.#.....#.#.....#.#....\n",
      "....#...#...#...#...#...#...#...#...................................#...#...#...#...#...#...#...#...\n",
      "...#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.................................#.#.#.#.#.#.#.#.#.#.#.#.#.#.#.#..\n",
      "..#...............................#...............................#...............................#.\n",
      ".#.#.............................#.#.............................#.#.............................#..\n",
      "Final state:  ....#...........................#...#...........................#...#...........................#.#.\n",
      "['.', '.', '.', '.', '#', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '#', '.', '.', '.', '#', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '#', '.', '.', '.', '#', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '.', '#', '.', '#', '.']\n"
     ]
    }
   ],
   "source": [
    "# Intended use case...\n",
    "class SilentCA(CellularAutomaton):\n",
    "    'A cellular automaton that does not print to the screen.'\n",
    "        \n",
    "    def process(self, time):\n",
    "        'Move one time step forward, do not print state.'\n",
    "        self.gridstate = self.applyrule(self.gridstate)\n",
    "    # ---\n",
    "# --- CellularAutomaton\n",
    "\n",
    "\n",
    "def simplelogging(entity):\n",
    "    'We will use this to print the state.'\n",
    "    print(entity)\n",
    "# ---\n",
    "\n",
    "\n",
    "## < --- --- ----\n",
    "\n",
    "\n",
    "sys = System()\n",
    "\n",
    "sys.add(SilentCA(100), \n",
    "        name='Sierpinsky')\n",
    "\n",
    "sys.run(steps=50, \n",
    "        before_step=[(simplelogging,['Sierpinsky'])] )\n",
    "\n",
    "print('Final state: ', sys['Sierpinsky'])\n",
    "print(sys.stateof('Sierpinsky'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from functools import wraps\n",
    "\n",
    "class Log:\n",
    "    \"A logger class that works as a register of the actions of a cell system.\"\n",
    "    \n",
    "    def __init__(self, *args, **kwargs):\n",
    "        pass\n",
    "    # ---\n",
    "    \n",
    "    def preparefor(self, actionname, *args, **kwargs):\n",
    "        'Save previous state before the entity takes the given action.'\n",
    "        getattr(self, 'preparefor_'+actionname)(*args, **kwargs)\n",
    "    # ---\n",
    "        \n",
    "    def log(self, actionname, *args, **kwargs):\n",
    "        'Log the action.'\n",
    "        getattr(self, 'log_'+actionname)(*args, **kwargs)\n",
    "    # ---\n",
    "# ---Log\n",
    "\n",
    "class SimpleLog(Log):\n",
    "    'Simple logger that limits to print the action.'\n",
    "    \n",
    "    def preparefor_division(self, cell):\n",
    "        print(\"Cell no. {} dividing @ {}\".format(cell.index,\n",
    "                                                 cell.coordinates))\n",
    "        \n",
    "    def log_division(self, daughters):\n",
    "        daughter1, daughter2 = daughters\n",
    "        print(\"\\tNew cells: {} @ {} and {} @ {}\".format(\n",
    "                    daughter1.index,\n",
    "                    daughter1.coordinates, \n",
    "                    daughter2.index, \n",
    "                    daughter2.coordinates))\n",
    "        \n",
    "    def preparefor_death(self, cell):\n",
    "        pass\n",
    "        \n",
    "    def log_death(self, cell):\n",
    "        print(\"Cell no. {} death @ site {} (father {})\".format(\n",
    "            cell.index, cell.coordinates, cell.father))\n",
    "        \n",
    "    def preparefor_migration(self, cell):\n",
    "        print(\"Cell no. {} migrating from site {} (father {})\".format(\n",
    "                cell.index, cell.coordinates, cell.father))\n",
    "        \n",
    "    def log_migration(self, cell):\n",
    "        print(\"\\t New site: {}\".format(cell.coordinates))\n",
    "        \n",
    "    def preparefor_mutation(self, cell):\n",
    "        print(\"Cell no. {} mutating @ site {} (father {})\".format(\n",
    "            cell.index, cell.coordinates, cell.father))\n",
    "        print(\"\\t\\t Initial mutations: {}\\n \\\n",
    "               \\t Initial genome: {}\".format(cell.mutations, cell.genome))\n",
    "        \n",
    "    def log_mutation(self, cell):\n",
    "        print(\"\\t\\t Final mutations: {}\\n \\\n",
    "               \\t Final genome: {}\".format(cell.mutations, cell.genome))\n",
    "        \n",
    "    def log_newcell(self, cell):\n",
    "        print( \"New cell {} added @ {}\".format(cell.index, cell.coordinates) )\n",
    "# --- SimpleLogger \n",
    "\n",
    "\n",
    "def logged(action_name, prepare=True):\n",
    "    \"\"\"Decorate an action with logging.\n",
    "    \n",
    "    Allows to specify if a prelogging action is called.\n",
    "    \"\"\"\n",
    "    # Real decorator\n",
    "    def add_logging(action):\n",
    "        # Decorated\n",
    "        @wraps(action)\n",
    "        def logged_action(self, *args, log=None, **kwargs):\n",
    "            if log and prepare:\n",
    "                log.preparefor(action_name, self)\n",
    "                \n",
    "            result = action(self, *args, **kwargs)\n",
    "            \n",
    "            if log:\n",
    "                log.log(action_name, result)\n",
    "        # ---\n",
    "        return logged_action\n",
    "    return add_logging\n",
    "# ---\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load cells.py\n",
    "\"\"\"The biological entities.\n",
    "\n",
    "Every cell belongs to a cell line.\n",
    "\n",
    "The cell line is in charge of:\n",
    "    - creating and initializing new cells\n",
    "    - managing the disposal of dead cells\n",
    "    - manage the ancestral genome of the descendants.\n",
    "    \n",
    "A cell can perform actions like:\n",
    "    - migrate, \n",
    "    - mutate,\n",
    "    - divide,\n",
    "    - die.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "import random as rnd\n",
    "\n",
    "\n",
    "class CellLine:\n",
    "    \"\"\"\n",
    "    Handles the specimens of a specific cell lineage.\n",
    "\n",
    "    A cell lineage is a group of cells that have a common ancestor, we\n",
    "    represent the common ancestor by it's ancestral genome.\n",
    "    This structure is in charge of holding this ancestral code and managing\n",
    "    the cells creating new cells when needed and cleaning up the dead ones.\n",
    "\n",
    "    Atributes:\n",
    "            + Ancestral genome: A string-like object.\n",
    "            + Cells: The cells inherited from this cell line.\n",
    "            + Alive/Dead cells.\n",
    "            + Current cell index: Each cell has a unique index. This is the\n",
    "                                  index to place in the next cell to be born.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self,\n",
    "                 system,\n",
    "                 genome=None,\n",
    "                 genomeAlphabet=None,\n",
    "                 recycleDeadCells=True):\n",
    "        \"\"\"Creation of a cell lineage.\n",
    "\n",
    "        Parameters:\n",
    "            :param genome: (optional) A sequence representing the base\n",
    "                            genome of the cells from this line.\n",
    "            :param genomeAlphabet: (optional) The items of which the genome\n",
    "                                   sequence is formed.\n",
    "\n",
    "            If a custom genome is passed, the genome alphabet should be\n",
    "            passed too unless it is formed of the letters in \"ACGT\".\n",
    "\n",
    "            Also, dead cells are recycled by default, this helps to improve\n",
    "            memory usage.\n",
    "\n",
    "        \"\"\"\n",
    "        self.recycleDeadCells = recycleDeadCells\n",
    "        self.currentIndex = 0\n",
    "        self.cells = []\n",
    "        self.aliveCells = set()\n",
    "        self.deadCells = set()\n",
    "        self.system = system\n",
    "\n",
    "        # Set the default genome\n",
    "        if not genome:\n",
    "            genome = 10 * 'A'\n",
    "        self.genome = genome\n",
    "\n",
    "        # Check if the genome alphabet option was set\n",
    "        if not genomeAlphabet:\n",
    "            genomeAlphabet = \"ACGT\"\n",
    "        self.genomeAlphabet = genomeAlphabet\n",
    "\n",
    "        # Validate the resulting genome/genome alphabet combination\n",
    "        if not self.validateGenome(genome, genomeAlphabet):\n",
    "            # ERROR\n",
    "            raise ValueError(\n",
    "                '\"genome\" must contain only letters\\\n",
    "                 present in the \"genomeAlphabet\"' )\n",
    "    # ---\n",
    "    \n",
    "    @classmethod\n",
    "    def validateGenome(cls, genome, genomeAlphabet):\n",
    "        \"\"\"Check if the genome is assembled from the genome alphabet.\"\"\"\n",
    "        # Eliminate dups\n",
    "        genomeLetters = set(genome)\n",
    "        # Are they valid?\n",
    "        return all( (l in genomeAlphabet) for l in genomeLetters )\n",
    "    # ---\n",
    "\n",
    "    @logged('newcell', prepare=False)\n",
    "    def addCellTo(self, site):\n",
    "        \"\"\"Add a new, initialized cell to the given site.\n",
    "\n",
    "        Return the added cell to the caller.\n",
    "\n",
    "        \"\"\"\n",
    "        new = self.newCell()\n",
    "        new.addTo(site)\n",
    "        # Return the new cell\n",
    "        return new\n",
    "    # ---\n",
    "\n",
    "    def newCell(self):\n",
    "        \"\"\"Get a new blank cell in this lineage and system.\"\"\"\n",
    "        # Fetch a blank cell\n",
    "        if self.recycleDeadCells and self.deadCells:\n",
    "            # Fetch a dead cell to recycle\n",
    "            new = self.cellToRecycle()\n",
    "            self.recycleCell(new)\n",
    "        else:\n",
    "            # Fetch a fresh, new cell\n",
    "            new = Cell( cellLine = self,\n",
    "                        index = self.currentIndex )\n",
    "            self.currentIndex += 1\n",
    "            self.cells.append(new)\n",
    "\n",
    "        # Update state to take new cell into account\n",
    "        self.aliveCells.add(new)\n",
    "        return new\n",
    "    # ---\n",
    "\n",
    "    def cellToRecycle(self):\n",
    "        \"\"\"Return a cell from the dead ones.\"\"\"\n",
    "        recycled = self.deadCells.pop()\n",
    "        return recycled\n",
    "    # ---\n",
    "\n",
    "    def recycleCell(self, cell):\n",
    "        \"\"\"Clear previous information from a cell.\"\"\"\n",
    "        # Place new ID\n",
    "        cell.index = self.currentIndex\n",
    "        self.currentIndex += 1\n",
    "        # Delete age\n",
    "        cell.age = 0\n",
    "    # ---\n",
    "\n",
    "    def getAliveCells(self):\n",
    "        \"\"\"Return the cells currently alive.\"\"\"\n",
    "        return self.aliveCells\n",
    "    # ---\n",
    "\n",
    "    def totalAliveCells(self):\n",
    "        \"\"\"Return the total number of alive cells.\"\"\"\n",
    "        return len(self.aliveCells)\n",
    "    # ---\n",
    "\n",
    "    def sample(self, all=False, n=1):\n",
    "        \"\"\"Take a sample of alive cells.\n",
    "\n",
    "        :param all: If True, return all alive cells, else, return a sample of\n",
    "                    size n.\n",
    "        :param n: The size of the sample. If 1, return the cell without a\n",
    "                  container.\n",
    "\n",
    "        \"\"\"\n",
    "        # If it is asked for all cells,\n",
    "        # take a sample of the whole size\n",
    "        if all:\n",
    "            n = len(self.aliveCells)\n",
    "        # If n equals 1, return a single cell\n",
    "        elif n == 1:\n",
    "            [singleCell] = rnd.sample(self.aliveCells, 1)\n",
    "            return singleCell\n",
    "        # Return a sample of size n\n",
    "        return rnd.sample(self.aliveCells, n)\n",
    "    # ---\n",
    "\n",
    "    def handleDeath(self, dying):\n",
    "        \"\"\"Process a dying cell.\n",
    "\n",
    "        This means removing from the alive cells and adding to the dead ones,\n",
    "        maybe to recicle it when another is born.\n",
    "\n",
    "        \"\"\"\n",
    "        self.aliveCells.remove(dying)\n",
    "        self.deadCells.add(dying)\n",
    "    # ---\n",
    "# --- CellLine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "class CellAction:\n",
    "    \"\"\"\n",
    "    Cell action.\n",
    "\n",
    "    Class used to represent actions that the cell could perform.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, action, actionProbability):\n",
    "        \"\"\"Initialize action object.\n",
    "\n",
    "        It is initialized with a function that when called will make\n",
    "        the cell perform the action and a function that when called will\n",
    "        calculate the probability to perform the action.\n",
    "\n",
    "\n",
    "        \"\"\"\n",
    "        self.action = action\n",
    "        self.actionProbability = actionProbability\n",
    "\n",
    "    def tryAction(self, log=None):\n",
    "        \"\"\"Perform the action according to it's probability.\"\"\"\n",
    "        # Sample according to the probability\n",
    "        if rnd.random() < self.actionProbability():\n",
    "            self.action(log = log)\n",
    "# --- CellAction\n",
    "\n",
    "\n",
    "class Cell:\n",
    "    \"\"\"\n",
    "    A single cell.\n",
    "\n",
    "    It acts according to it's state, and the states of nearby cells and sites.\n",
    "\n",
    "    Attributes:\n",
    "        + Index: A label that identifies it among others in the\n",
    "                 same lineage.\n",
    "        + Father: The index (lineage label) of it's father.\n",
    "        + CellLine: The lineage this cell belongs to.\n",
    "        + Site: The place in the grid this cell inhabits in.\n",
    "        + Age: The timesteps this cell has passed through.\n",
    "        + Mutations: The mutations in this cell relative to the cell\n",
    "                     lineage's reference\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, cellLine, index):\n",
    "        \"\"\"Create a new cell.\n",
    "\n",
    "        Parameters:\n",
    "            :param cellLine: The lineage this cell belongs to.\n",
    "            :param index: The ID of this cell in the lineage.\n",
    "\n",
    "        \"\"\"\n",
    "        self.index = index\n",
    "        self.cellLine = cellLine\n",
    "        self.father = None\n",
    "        self.site = None\n",
    "        self.age = 0\n",
    "        self.availableActions = self.initializeActions()\n",
    "        self.mutations = []\n",
    "    # ---\n",
    "\n",
    "    @property\n",
    "    def coordinates(self):\n",
    "        \"\"\"Coordinates of the site that the cell inhabits.\"\"\"\n",
    "        return self.site.coordinates\n",
    "\n",
    "    @coordinates.setter\n",
    "    def coordinates(self, value):\n",
    "        \"\"\"Set site by coordinates.\"\"\"\n",
    "        self.site = self.system.at(*value)\n",
    "    # ---\n",
    "\n",
    "    def resetAge(self):\n",
    "        \"\"\"Return the cell to it's infantry.\"\"\"\n",
    "        self.age = 0\n",
    "    # ---\n",
    "\n",
    "    def olderThan(self, age):\n",
    "        \"\"\"Is this cell older than that?\"\"\"\n",
    "        return self.age > age\n",
    "    # ---\n",
    "    \n",
    "    @property\n",
    "    def mutations(self):\n",
    "        \"\"\"Record of the mutations the cell has had.\n",
    "\n",
    "        The mutations are relative to the ancestral genome.\n",
    "\n",
    "        \"\"\"\n",
    "        return self._mutations\n",
    "\n",
    "    @mutations.setter\n",
    "    def mutations(self, value):\n",
    "        \"\"\"Setter for the cell's mutations.\"\"\"\n",
    "        self._mutations = list(value)  # COPY\n",
    "    # ---\n",
    "\n",
    "    @property\n",
    "    def ancestralGenome(self):\n",
    "        \"\"\"Ancestral genome of the cell lineage.\"\"\"\n",
    "        return self.cellLine.genome\n",
    "    # ---\n",
    "\n",
    "    @property\n",
    "    def genome(self):\n",
    "        \"\"\"Genome of the cell.\n",
    "\n",
    "        It is assembled from the cell's ancestral genome and it's mutations.\n",
    "\n",
    "        \"\"\"\n",
    "        # Get the original genome\n",
    "        bases = list(self.ancestralGenome)\n",
    "        # Add mutations one by one\n",
    "        for pos, mutated in self.mutations:\n",
    "            bases[pos] = mutated\n",
    "        # Assemble the final genome\n",
    "        genome = ''.join(bases)\n",
    "\n",
    "        return genome\n",
    "    # ---\n",
    "\n",
    "    def initializeWith(self, site=None, father=None, mutations=None):\n",
    "        \"\"\"Initialize grid site, mutations and father.\"\"\"\n",
    "        if site:\n",
    "            self.site = site\n",
    "        if father:\n",
    "            self.father = father\n",
    "        if mutations:\n",
    "            self.mutations = mutations  # copy\n",
    "        self.age = 0\n",
    "    # ---\n",
    "\n",
    "    @logged('division')\n",
    "    def divide(self, preserveFather=False):\n",
    "        \"\"\"Cell division.\n",
    "\n",
    "        Get a new daughter of this cell and place it in a nearby\n",
    "        neighboring site.\n",
    "\n",
    "        \"\"\"\n",
    "        # Create the daughter cell\n",
    "        daughter = self.newDaughter()\n",
    "        # Search for the site to place the daughter in\n",
    "        site = self.site.randomNeighbor()\n",
    "        daughter.addTo(site)\n",
    "\n",
    "        # If the preserveFather flag is set, we want a\n",
    "        # father cell and a daughter cell after the division,\n",
    "        # else, we want both resulting cells to be daughters of the\n",
    "        # father cell, so this cell will migrate, set to age 0 and\n",
    "        # migrate.\n",
    "        if not preserveFather:\n",
    "            self.resetAge()\n",
    "            self.father = self.index\n",
    "            self.cellLine.recycleCell(self)\n",
    "            self.migrate()\n",
    "            \n",
    "        return self, daughter\n",
    "    # ---\n",
    "\n",
    "    def divisionProbability(self):\n",
    "        \"\"\"Probability that this cell will divide if selected for division.\"\"\"\n",
    "        return 1\n",
    "    # ---\n",
    "\n",
    "    def newDaughter(self):\n",
    "        \"\"\"Initialize a new daughter cell.\n",
    "\n",
    "        Initialize with father and mutation attributes.\n",
    "\n",
    "        \"\"\"\n",
    "        daughter = self.cellLine.newCell()\n",
    "        daughter.initializeWith( father=self.index,\n",
    "                                 mutations=self.mutations )\n",
    "        return daughter\n",
    "    # ---\n",
    "\n",
    "    def growOlder(self):\n",
    "        \"\"\"Increase the age of the current cell.\"\"\"\n",
    "        self.age += 1\n",
    "    # ---\n",
    "\n",
    "    def addTo(self, site):\n",
    "        \"\"\"Add the cell to the site.\"\"\"\n",
    "        self.site = site\n",
    "        site.addGuest(self)\n",
    "    # ---\n",
    "\n",
    "    @logged('death')\n",
    "    def die(self):\n",
    "        \"\"\"Cellular death.\"\"\"\n",
    "        self.site.removeGuest(self)\n",
    "        self.cellLine.handleDeath(self)\n",
    "        return self\n",
    "    # ---\n",
    "\n",
    "    def deathProbability(self):\n",
    "        \"\"\"Cellular death probability.\"\"\"\n",
    "        aliveCells = self.cellLine.totalAliveCells()\n",
    "        if aliveCells > 1:\n",
    "            return 0.6\n",
    "        else:\n",
    "            return 0\n",
    "    # ---\n",
    "    \n",
    "    @logged('migration')\n",
    "    def migrate(self):\n",
    "        \"\"\"Migrate to a neighboring cell.\"\"\"\n",
    "        # Get the destination site\n",
    "        nextSite = self.site.randomNeighbor()\n",
    "        # Migrate to the new site\n",
    "        self.site.removeGuest(self)\n",
    "        nextSite.addGuest(self)\n",
    "        self.site = nextSite\n",
    "        \n",
    "        return self\n",
    "    # ---\n",
    "\n",
    "    def migrationProbability(self):\n",
    "        \"\"\"Migration probability for this cell.\"\"\"\n",
    "        return 1\n",
    "\n",
    "    # ---\n",
    "    \n",
    "    @logged('mutation')\n",
    "    def mutate(self):\n",
    "        \"\"\"Do a single site mutation.\n",
    "\n",
    "        Note: The genome may not represent a nucleotide sequence, so these\n",
    "        mutations may not represent SNPs.\n",
    "\n",
    "        \"\"\"\n",
    "        # Get the genome characteristics\n",
    "        # Convert alphabet to tuple to call rnd.choice with it\n",
    "        alphabet = tuple(self.cellLine.genomeAlphabet)\n",
    "        genomeLength = len(self.genome)\n",
    "        # Assemble the mutation\n",
    "        position = rnd.randint(0, \n",
    "                               genomeLength - 1)  # Pick a random position in the genome\n",
    "        mutated = rnd.choice(alphabet)\n",
    "        # Mutate\n",
    "        self.mutations.append((position, mutated))\n",
    "        \n",
    "        return self\n",
    "    # ---\n",
    "\n",
    "    def mutationProbability(self):\n",
    "        \"\"\"Probability to mutate if selected for it.\"\"\"\n",
    "        return 1\n",
    "    # ---\n",
    "\n",
    "    def initializeActions(self):\n",
    "        \"\"\"Return a list with the possible actions to take for this cell.\"\"\"\n",
    "        # Death action\n",
    "        death = CellAction(self.die, self.deathProbability)\n",
    "        # Division action\n",
    "        division = CellAction(self.divide, self.divisionProbability)\n",
    "\n",
    "        # Migration action\n",
    "        migration = CellAction(self.migrate, self.migrationProbability)\n",
    "\n",
    "        # Mutation action\n",
    "        mutation = CellAction(self.mutate, self.mutationProbability)\n",
    "\n",
    "        return [mutation, migration, division, death]\n",
    "    # ---\n",
    "\n",
    "    def step(self, log=None):\n",
    "        \"\"\"Select an action and perform it.\"\"\"\n",
    "        # Make the cell older\n",
    "        self.growOlder()\n",
    "        # Select an action\n",
    "        action = rnd.choice(self.availableActions)\n",
    "        # Perform the action according to it's respective probability\n",
    "        action.tryAction(log=log)\n",
    "    # ---\n",
    "# --- Cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load system.py\n",
    "\"\"\"System related classes.\n",
    "\n",
    "This module handles the classes Site and System. Related to the non-biological\n",
    "part of the simulation.\n",
    "\"\"\"\n",
    "\n",
    "import random as rnd\n",
    "\n",
    "\n",
    "class Site:\n",
    "    \"\"\"\n",
    "    A unit of space.\n",
    "\n",
    "    Cells inhabitate in these spaces and interact with their neighborhood.\n",
    "\n",
    "    Is aware of:\n",
    "            + System: The system it forms a part of.\n",
    "            + Coordinates <i,j>: Coordinates in the matrix.\n",
    "            + Guests: <List of cells>: The cells currently in this site.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, system, i, j):\n",
    "        \"\"\"Assemble a site in which agents may inhabit.\n",
    "\n",
    "        :param system: The system in which this system resides.\n",
    "        :param i: Vertical position.\n",
    "        :param j: Horizontal position.\n",
    "\n",
    "        \"\"\"\n",
    "        self.system = system\n",
    "        self._coordinates = (i, j)  # Set variable only once\n",
    "        self.guests = set()\n",
    "    # ---\n",
    "\n",
    "    @property\n",
    "    def coordinates(self):\n",
    "        \"\"\"Getter for the site's coordinates.\"\"\"\n",
    "        return self._coordinates\n",
    "    # ---\n",
    "\n",
    "    def addGuest(self, guest):\n",
    "        \"\"\"Add the given cell as a new guest to this site.\"\"\"\n",
    "        self.guests.add(guest)\n",
    "        guest.site = self\n",
    "    # ---\n",
    "\n",
    "    def removeGuest(self, guest):\n",
    "        \"\"\"Remove the given cell as guest for this site.\n",
    "\n",
    "        If the cell is not currently in this site, an error is throwed.\n",
    "        \n",
    "        \"\"\"\n",
    "        try:\n",
    "            self.guests.remove(guest)\n",
    "        except KeyError:\n",
    "            raise KeyError(\n",
    "                    'Guest with index {} is not at site ({},{})'\n",
    "                    .format(guest.index, i, j)\n",
    "                  )\n",
    "    # ---\n",
    "\n",
    "    def guestCount(self):\n",
    "        \"\"\"Return the number of guests residing in this site.\"\"\"\n",
    "        return len(self.guests)\n",
    "    # ---\n",
    "\n",
    "    def randomNeighbor(self):\n",
    "        \"\"\"Return a random site in the neighborhood of the current site.\"\"\"\n",
    "        # Get your actual coordinates\n",
    "        i, j = self.coordinates\n",
    "        # Select a neighbor\n",
    "        neighborhood = self.system.getNeighborhood()\n",
    "        n_i, n_j = rnd.choice(neighborhood)\n",
    "        # Fetch the neighbor\n",
    "        return self.system.at( i+n_i, j+n_j )\n",
    "    # ---\n",
    "\n",
    "\n",
    "# < ---------------------------------------------------------------------------\n",
    "\n",
    "def wrap(n, maxValue):\n",
    "    \"\"\"Auxiliary function to wrap an integer on maxValue.\n",
    "\n",
    "    Examples:\n",
    "            >>> # For positives: wrap(n, maxValue) = n % maxValue\n",
    "            >>> [ wrap(i,3) for i in range(9) ]\n",
    "            [0, 1, 2, 0, 1, 2, 0, 1, 2]\n",
    "\n",
    "            >>> # For negatives, the pattern is continued in a natural way\n",
    "            >>> for i in range(-5, 5+1): print(f'{i} : {wrap(i, 3)}')\n",
    "            ...\n",
    "            -3 : 0\n",
    "            -2 : 1\n",
    "            -1 : 2\n",
    "            0 : 0\n",
    "            1 : 1\n",
    "            2 : 2\n",
    "\n",
    "    \"\"\"\n",
    "    if n >= maxValue:\n",
    "        n %= maxValue\n",
    "    elif n < 0:\n",
    "        # We must offset the integer to the positives\n",
    "        n += abs(n//maxValue)*maxValue\n",
    "    return n\n",
    "# ---\n",
    "\n",
    "\n",
    "# < -----\n",
    "class CellSystem:\n",
    "    \"\"\"\n",
    "    The global system and event dispatcher.\n",
    "\n",
    "    Aware of cells, space (as a grid) and the passage of time (steps). It is\n",
    "    the main struture of the program, being the only observable and given it\n",
    "    coordinates all processes.\n",
    "\n",
    "    A system is aware of:\n",
    "            - Grid: The sites the action develops in.\n",
    "            - Cells: The cells that live, die and interact in the grid.\n",
    "            - Neighborhood: How many and which sites may directly influence or\n",
    "                            be influenced by another.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, gridDimensions=(10, 10), genome=None, wrap=True):\n",
    "        \"\"\"Initialize the system.\n",
    "\n",
    "        :param gridDimensions: Tuple specifying rows and columns.\n",
    "        :param genome: Container specifying the allowed 'letters' for a genome.\n",
    "        :param wrap: Boolean. Does the grid wraps on the edges? (toroidal)\n",
    "\n",
    "        \"\"\"\n",
    "        # Set dimensions\n",
    "        self.rows, self.cols = gridDimensions\n",
    "        # Initialize grid\n",
    "        self.wrap = wrap  # Toroidal wrapping behavior of the grid\n",
    "        self.grid = tuple( Site(self, i,j)\n",
    "                               for i in range(self.rows)\n",
    "                                   for j in range(self.cols) )\n",
    "        # Initialize cells\n",
    "        self.cells = CellLine(genome = genome,\n",
    "                              system = self)\n",
    "        # Initialize neighborhood\n",
    "        self.neighborhood = [ (-1,-1), (-1, 0), (-1, 1),\n",
    "                              ( 0,-1), ( 0, 0), ( 0, 1),\n",
    "                              ( 1,-1), ( 1, 0), ( 1, 1) ]\n",
    "    # ---\n",
    "\n",
    "    def seed(self, at_coords=None, log=None):\n",
    "        \"\"\"Seed automaton with a cell at the given coordinates.\n",
    "\n",
    "        If explicit coordinates are not given, place a single cell in the\n",
    "        middle of the grid.\n",
    "\n",
    "        \"\"\"\n",
    "        if at_coords is None:\n",
    "            # Seed in the middle of the grid\n",
    "            at_coords = self.rows//2, self.cols//2\n",
    "        # Get the site we're adding the cell to\n",
    "        site = self.at(*at_coords)\n",
    "        # Add a new cell\n",
    "        self.cells.addCellTo(site, log=log)\n",
    "    # ---\n",
    "\n",
    "    def cellCountAt(self, i, j):\n",
    "        \"\"\"Return the number of cells in a given site.\"\"\"\n",
    "        return self.at(i,j).guestCount()\n",
    "    # ---\n",
    "\n",
    "    def step(self, steps=1, singleCell=False, log=None):\n",
    "        \"\"\"Move the state of the system `steps` steps forward in time.\"\"\"\n",
    "        for _ in range(steps):\n",
    "            self.singleStep(singleCell, log=log)\n",
    "    # ---\n",
    "\n",
    "    def singleStep(self, singleCell=False, log=None):\n",
    "        \"\"\"Move the state of the system one step forward in time.\"\"\"\n",
    "        # Advance a single cell\n",
    "        if singleCell:\n",
    "            cell = self.cells.sample()\n",
    "            cell.step(log=log)\n",
    "        else:\n",
    "            # Advance all cells\n",
    "            for cell in self.cells.sample(all=True):\n",
    "                # Perform an action on the cell\n",
    "                cell.step(log=log)\n",
    "    # ---\n",
    "\n",
    "    def at(self, i, j):\n",
    "        \"\"\"Get the site at the specified coordinates.\"\"\"\n",
    "        # Wrap (toroidal coordinates)\n",
    "        if self.wrap:\n",
    "            i,j = self.wrapCoordinates(i,j)\n",
    "        # The grid is 1D, so we must convert from 2D\n",
    "        ij = self.cols*i + j\n",
    "        return self.grid[ij]\n",
    "    # ---\n",
    "\n",
    "    def wrapCoordinates(self, i, j):\n",
    "        \"\"\"Return i,j wrapped on the grid dimensions.\"\"\"\n",
    "        # Wrap the coordinates\n",
    "        i = wrap(i, self.rows)\n",
    "        j = wrap(j, self.cols)\n",
    "        return i,j\n",
    "\n",
    "    def totalCells(self, state='alive'):\n",
    "        \"\"\"Return the total number of cells of with the given 'state'.\"\"\"\n",
    "        if state == 'alive':\n",
    "            # Total count of alive cells\n",
    "            return self.cells.totalAliveCells()\n",
    "        elif state == 'dead':\n",
    "            # Total count of dead cells\n",
    "            # (makes sense only if the recycle dead cells option is not set)\n",
    "            return self.cells.totalDeadCells()\n",
    "        elif state == 'all':\n",
    "            # Total count of cells in the system\n",
    "            # (this is usually for debug purposes)\n",
    "            return self.cells.totalCells()\n",
    "    # ---\n",
    "\n",
    "    def totalAliveCells(self):\n",
    "        \"\"\"Return the total number of cells in the system.\"\"\"\n",
    "        return self.cells.totalAliveCells()\n",
    "    # ---\n",
    "\n",
    "    def getNeighborhood(self):\n",
    "        \"\"\"Return the relative coordinates of the available neighbors.\n",
    "\n",
    "        Returns a container holding pairs of numbers. the prescence of an item\n",
    "        (a,b) means that a site at i,j has a neighbor at (i+a, j+b).\n",
    "\n",
    "        \"\"\"\n",
    "        return self.neighborhood\n",
    "    # ---\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys = CellSystem()\n",
    "log = SimpleLog()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "New cell 0 added @ (5, 5)\n"
     ]
    }
   ],
   "source": [
    "sys.seed(log=log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cell no. 0 migrating from site (5, 5) (father None)\n",
      "\t New site: (4, 5)\n",
      "Cell no. 0 mutating @ site (4, 5) (father None)\n",
      "\t\t Initial mutations: []\n",
      "                \t Initial genome: AAAAAAAAAA\n",
      "\t\t Final mutations: [(7, 'T')]\n",
      "                \t Final genome: AAAAAAATAA\n",
      "Cell no. 0 migrating from site (4, 5) (father None)\n",
      "\t New site: (4, 4)\n",
      "Cell no. 0 mutating @ site (4, 4) (father None)\n",
      "\t\t Initial mutations: [(7, 'T')]\n",
      "                \t Initial genome: AAAAAAATAA\n",
      "\t\t Final mutations: [(7, 'T'), (1, 'G')]\n",
      "                \t Final genome: AGAAAAATAA\n",
      "Cell no. 0 mutating @ site (4, 4) (father None)\n",
      "\t\t Initial mutations: [(7, 'T'), (1, 'G')]\n",
      "                \t Initial genome: AGAAAAATAA\n",
      "\t\t Final mutations: [(7, 'T'), (1, 'G'), (9, 'G')]\n",
      "                \t Final genome: AGAAAAATAG\n",
      "Cell no. 0 dividing @ (4, 4)\n",
      "\tNew cells: 2 @ (4, 5) and 1 @ (5, 5)\n",
      "Cell no. 2 dividing @ (4, 5)\n",
      "\tNew cells: 4 @ (3, 4) and 3 @ (3, 5)\n",
      "Cell no. 1 dividing @ (5, 5)\n",
      "\tNew cells: 6 @ (6, 4) and 5 @ (5, 5)\n",
      "Cell no. 6 mutating @ site (6, 4) (father 1)\n",
      "\t\t Initial mutations: [(7, 'T'), (1, 'G'), (9, 'G')]\n",
      "                \t Initial genome: AGAAAAATAG\n",
      "\t\t Final mutations: [(7, 'T'), (1, 'G'), (9, 'G'), (8, 'C')]\n",
      "                \t Final genome: AGAAAAATCG\n",
      "Cell no. 4 dividing @ (3, 4)\n",
      "\tNew cells: 8 @ (2, 4) and 7 @ (4, 3)\n",
      "Cell no. 3 dividing @ (3, 5)\n",
      "\tNew cells: 10 @ (4, 5) and 9 @ (4, 6)\n",
      "Cell no. 5 migrating from site (5, 5) (father 1)\n",
      "\t New site: (5, 4)\n",
      "Cell no. 7 mutating @ site (4, 3) (father 4)\n",
      "\t\t Initial mutations: [(7, 'T'), (1, 'G'), (9, 'G')]\n",
      "                \t Initial genome: AGAAAAATAG\n",
      "\t\t Final mutations: [(7, 'T'), (1, 'G'), (9, 'G'), (9, 'T')]\n",
      "                \t Final genome: AGAAAAATAT\n",
      "Cell no. 9 mutating @ site (4, 6) (father 3)\n",
      "\t\t Initial mutations: [(7, 'T'), (1, 'G'), (9, 'G')]\n",
      "                \t Initial genome: AGAAAAATAG\n",
      "\t\t Final mutations: [(7, 'T'), (1, 'G'), (9, 'G'), (6, 'G')]\n",
      "                \t Final genome: AGAAAAGTAG\n",
      "Cell no. 6 death @ site (6, 4) (father 1)\n",
      "Cell no. 5 migrating from site (5, 4) (father 1)\n",
      "\t New site: (4, 4)\n",
      "Cell no. 8 migrating from site (2, 4) (father 4)\n",
      "\t New site: (2, 5)\n",
      "Cell no. 10 mutating @ site (4, 5) (father 3)\n",
      "\t\t Initial mutations: [(7, 'T'), (1, 'G'), (9, 'G')]\n",
      "                \t Initial genome: AGAAAAATAG\n",
      "\t\t Final mutations: [(7, 'T'), (1, 'G'), (9, 'G'), (6, 'T')]\n",
      "                \t Final genome: AGAAAATTAG\n"
     ]
    }
   ],
   "source": [
    "sys.step(10, log=log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load log.py\n",
    "import sys\n",
    "        \n",
    "from ete3 import Tree\n",
    "\n",
    "\n",
    "class WeakLog(Log):\n",
    "    'Ignore silently calls to not implemented log methods.'\n",
    "    \n",
    "    def preparefor(self, action, *args, **kwargs):\n",
    "        try:\n",
    "            logaction = getattr(self, 'preparefor_'+action)\n",
    "        except AttributeError:\n",
    "            return\n",
    "        else:\n",
    "            logaction(*args, **kwargs)\n",
    "            \n",
    "    def log(self, action, *args, **kwargs):\n",
    "        try:\n",
    "            logaction = getattr(self, 'log_'+action)\n",
    "        except AttributeError:\n",
    "            return\n",
    "        else:\n",
    "            logaction(*args, **kwargs)\n",
    "    \n",
    "\n",
    "class AncestryLog(WeakLog):\n",
    "    \n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super().__init__(*args, **kwargs)\n",
    "        self.ancestry = Tree()\n",
    "        self.alive = {}\n",
    "        self.tmp = None\n",
    "    \n",
    "    def log_newcell(self, cell):\n",
    "        self.alive[cell.index] = self.ancestry\\\n",
    "                                    .add_child(name=cell.index)\n",
    "        \n",
    "    def preparefor_division(self, cell):\n",
    "        self.tmp = cell.index\n",
    "        \n",
    "    def log_division(self, daughters):\n",
    "        d1, d2 = daughters\n",
    "        father = self.tmp\n",
    "        self.alive[d1.index] = self.alive[father]\\\n",
    "                                .add_child(name=d1.index)\n",
    "        self.alive[d2.index] = self.alive[father]\\\n",
    "                                .add_child(name=d2.index)\n",
    "        del self.alive[father]\n",
    "        self.tmp = None\n",
    "        \n",
    "    def log_death(self, cell):\n",
    "        # No further childs, no need to track\n",
    "        del self.alive[cell.index]        \n",
    "        \n",
    "     \n",
    "class MutationsLog(WeakLog):\n",
    "    \n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super().__init__(*args, **kwargs)\n",
    "        self.mutations = Tree()\n",
    "        self.alive = {}\n",
    "        self.tmp = None\n",
    "        \n",
    "    def log_newcell(self, cell):\n",
    "        self.alive[cell.index] = self.mutations\\\n",
    "                                    .add_child(name=cell.index)\n",
    "        \n",
    "    def preparefor_division(self, cell):\n",
    "        self.tmp = cell.index\n",
    "        \n",
    "    def log_division(self, daughters):\n",
    "        d1, d2 = daughters\n",
    "        father = self.tmp\n",
    "        node = self.alive[father]\n",
    "        self.alive[d1.index] = node\n",
    "        self.alive[d2.index] = node\n",
    "        # Cleanup\n",
    "        del self.alive[father]\n",
    "        self.tmp = None\n",
    "        \n",
    "    def log_death(self, cell):\n",
    "        # No further mutations, so no need to hold it\n",
    "        del self.alive[cell.index]\n",
    "        \n",
    "    def log_mutation(self, cell):\n",
    "        node = self.alive[cell.index]\n",
    "        new = node.add_child(name=cell.genome)\n",
    "        self.alive[cell.index] = new\n",
    "        \n",
    "\n",
    "class MALog(Log):\n",
    "    \n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super().__init__(*args, **kwargs)\n",
    "        self.mutations = MutationsLog()\n",
    "        self.ancestry = AncestryLog()\n",
    "        \n",
    "    def preparefor(self, actionname, *args, **kwargs):\n",
    "        'Save previous state before the entity takes the given action.'\n",
    "        self.mutations.preparefor(actionname, *args, **kwargs)\n",
    "        self.ancestry.preparefor(actionname, *args, **kwargs)\n",
    "    # ---\n",
    "        \n",
    "    def log(self, actionname, *args, **kwargs):\n",
    "        'Log the action.'\n",
    "        self.mutations.log(actionname, *args, **kwargs)\n",
    "        self.ancestry.log(actionname, *args, **kwargs)\n",
    "    # ---\n",
    "\n",
    "    \n",
    "class FullLog(Log):\n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super().__init__(*args, **kwargs)\n",
    "        self.mutationslog = MutationsLog()\n",
    "        self.ancestrylog = AncestryLog()\n",
    "        self.simplelog = SimpleLog()\n",
    "        \n",
    "    def preparefor(self, actionname, *args, **kwargs):\n",
    "        'Save previous state before the entity takes the given action.'\n",
    "        self.mutationslog.preparefor(actionname, *args, **kwargs)\n",
    "        self.ancestrylog.preparefor(actionname, *args, **kwargs)\n",
    "        self.simplelog.preparefor(actionname, *args, **kwargs)\n",
    "    # ---\n",
    "        \n",
    "    def log(self, actionname, *args, **kwargs):\n",
    "        'Log the action.'\n",
    "        self.mutationslog.log(actionname, *args, **kwargs)\n",
    "        self.ancestrylog.log(actionname, *args, **kwargs)\n",
    "        self.simplelog.log(actionname, *args, **kwargs)\n",
    "    # ---\n",
    "    \n",
    "    def ancestry(self, prune_death=False):\n",
    "        t = self.ancestrylog.ancestry\n",
    "        \n",
    "        if prune_death:\n",
    "            alive_nodes = {k for k in log.ancestrylog.alive.keys()}\n",
    "            return self.prune_death(t, alive_nodes)\n",
    "        \n",
    "        return t\n",
    "    \n",
    "    def mutations(self, prune_death=False):\n",
    "        t = self.mutationslog.mutations\n",
    "        \n",
    "        if prune_death:\n",
    "            alive_nodes = {v.name for v in log.mutationslog.alive.values()}\n",
    "            return self.prune_death(t, alive_nodes)\n",
    "        \n",
    "        return t\n",
    "    \n",
    "    @staticmethod\n",
    "    def prune_death(tree, alive_nodes):\n",
    "        # Prunning tree to represent only alive cells \n",
    "        t = tree.copy()\n",
    "        ancestors = set([t])\n",
    "        for node in t.traverse():\n",
    "            if node.name in alive_nodes:\n",
    "                ancestors |= set(node.get_descendants())\n",
    "                ancestors.add(node)\n",
    "        t.prune(ancestors)\n",
    "        return t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys = CellSystem()\n",
    "log = FullLog()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "New cell 0 added @ (5, 5)\n"
     ]
    }
   ],
   "source": [
    "sys.seed(log=log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cell no. 0 mutating @ site (5, 5) (father None)\n",
      "\t\t Initial mutations: []\n",
      "                \t Initial genome: AAAAAAAAAA\n",
      "\t\t Final mutations: [(2, 'A')]\n",
      "                \t Final genome: AAAAAAAAAA\n",
      "Cell no. 0 mutating @ site (5, 5) (father None)\n",
      "\t\t Initial mutations: [(2, 'A')]\n",
      "                \t Initial genome: AAAAAAAAAA\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C')]\n",
      "                \t Final genome: AAAAAAAACA\n",
      "Cell no. 0 mutating @ site (5, 5) (father None)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C')]\n",
      "                \t Initial genome: AAAAAAAACA\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T')]\n",
      "                \t Final genome: AAATAAAACA\n",
      "Cell no. 0 dividing @ (5, 5)\n",
      "\tNew cells: 2 @ (6, 4) and 1 @ (6, 5)\n",
      "Cell no. 1 mutating @ site (6, 5) (father None)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C'), (3, 'T')]\n",
      "                \t Initial genome: AAATAAAACA\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A')]\n",
      "                \t Final genome: AAATAAAAAA\n",
      "Cell no. 2 dividing @ (6, 4)\n",
      "\tNew cells: 4 @ (7, 3) and 3 @ (6, 5)\n",
      "Cell no. 1 mutating @ site (6, 5) (father None)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A')]\n",
      "                \t Initial genome: AAATAAAAAA\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C')]\n",
      "                \t Final genome: AAATAAAAAC\n",
      "Cell no. 4 dividing @ (7, 3)\n",
      "\tNew cells: 6 @ (8, 3) and 5 @ (6, 3)\n",
      "Cell no. 1 mutating @ site (6, 5) (father None)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C')]\n",
      "                \t Initial genome: AAATAAAAAC\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C')]\n",
      "                \t Final genome: AAATAAAAAC\n",
      "Cell no. 3 mutating @ site (6, 5) (father 2)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C'), (3, 'T')]\n",
      "                \t Initial genome: AAATAAAACA\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (5, 'G')]\n",
      "                \t Final genome: AAATAGAACA\n",
      "Cell no. 6 death @ site (8, 3) (father 4)\n",
      "Cell no. 5 mutating @ site (6, 3) (father 4)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C'), (3, 'T')]\n",
      "                \t Initial genome: AAATAAAACA\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (9, 'G')]\n",
      "                \t Final genome: AAATAAAACG\n",
      "Cell no. 3 death @ site (6, 5) (father 2)\n",
      "Cell no. 5 dividing @ (6, 3)\n",
      "\tNew cells: 8 @ (7, 3) and 7 @ (5, 4)\n",
      "Cell no. 1 dividing @ (6, 5)\n",
      "\tNew cells: 10 @ (6, 5) and 9 @ (7, 4)\n",
      "Cell no. 7 migrating from site (5, 4) (father 5)\n",
      "\t New site: (5, 3)\n",
      "Cell no. 10 mutating @ site (6, 5) (father 1)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C')]\n",
      "                \t Initial genome: AAATAAAAAC\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C'), (0, 'C')]\n",
      "                \t Final genome: CAATAAAAAC\n",
      "Cell no. 8 death @ site (7, 3) (father 5)\n",
      "Cell no. 10 death @ site (6, 5) (father 1)\n",
      "Cell no. 7 migrating from site (5, 3) (father 5)\n",
      "\t New site: (4, 3)\n",
      "Cell no. 9 mutating @ site (7, 4) (father 1)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C')]\n",
      "                \t Initial genome: AAATAAAAAC\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C'), (8, 'T')]\n",
      "                \t Final genome: AAATAAAATC\n",
      "Cell no. 7 migrating from site (4, 3) (father 5)\n",
      "\t New site: (3, 3)\n",
      "Cell no. 9 migrating from site (7, 4) (father 1)\n",
      "\t New site: (7, 3)\n",
      "Cell no. 7 death @ site (3, 3) (father 5)\n",
      "Cell no. 9 mutating @ site (7, 3) (father 1)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C'), (8, 'T')]\n",
      "                \t Initial genome: AAATAAAATC\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C'), (8, 'T'), (6, 'G')]\n",
      "                \t Final genome: AAATAAGATC\n",
      "Cell no. 9 mutating @ site (7, 3) (father 1)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C'), (8, 'T'), (6, 'G')]\n",
      "                \t Initial genome: AAATAAGATC\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C'), (8, 'T'), (6, 'G'), (4, 'T')]\n",
      "                \t Final genome: AAATTAGATC\n",
      "Cell no. 9 mutating @ site (7, 3) (father 1)\n",
      "\t\t Initial mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C'), (8, 'T'), (6, 'G'), (4, 'T')]\n",
      "                \t Initial genome: AAATTAGATC\n",
      "\t\t Final mutations: [(2, 'A'), (8, 'C'), (3, 'T'), (8, 'A'), (9, 'C'), (9, 'C'), (8, 'T'), (6, 'G'), (4, 'T'), (9, 'A')]\n",
      "                \t Final genome: AAATTAGATA\n",
      "Cell no. 9 migrating from site (7, 3) (father 1)\n",
      "\t New site: (8, 4)\n",
      "Cell no. 9 migrating from site (8, 4) (father 1)\n",
      "\t New site: (8, 4)\n",
      "Cell no. 9 dividing @ (8, 4)\n",
      "\tNew cells: 12 @ (8, 5) and 11 @ (8, 5)\n"
     ]
    }
   ],
   "source": [
    "sys.step(20, log=log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAPQAAACdCAYAAACD1FVsAAAACXBIWXMAAA3XAAAN1wFCKJt4AAAR7ElEQVR4nO3de1DU9f7H8Sc3Xa6KohuKwknDUAE1AR018NYh8oKm/AjLSzA/pclLljnzqzPE7zSNc7qMVkcnU386VCLqT800jZogTx4vPw1F8oZ35EBeUNwFXFk+vz+MPS7eIfnufnk/ZvjD/eyOr92ZF5/dL/v+fl2UUgohhC64ah1ACPHHkUILoSNSaCF0RAotRAOVlZVMmjQJf39/OnbsyFtvvYWzHGqSQgvRQHp6Om5ubpSWlvLLL7+Ql5dHQUGB1rEeiIsc5RYtnanKyopN5aQmGrHUVBIYGMjZs2fp2LGj1tEemuzQokUzVVmJ+I99/OXvpwifuI+9+w5gNBpZvnw5oaGhhIWF8dFHH2kd84G5ax1AtCz5+fnExcVpHePfOkzAL3QOlWYr4ELO1pOUlpZiMBg4evQohw8fZtiwYXTr1o2xY8dqnfa+ZIcWzSovL4+MjAyUUg7xYzqTjRsm/LzdCWjjRlLCn3B3d2f27Nm4uLjQs2dPXnjhBb799lutX7oHIju0aNG8Pd0o+X4kS9eXMn1CJy5dKMVisWAymfDz8wNAKYWrq3Psfc6RUohHyMvgypxJQXi2diUoKIj4+HjeeustamtrKS4uJjs7m+eee07rmA9ECi1EA59//jnFxcUEBAQwcuRI5s+f7zSFlrfcQjTQqVMnp/nM3JDs0ELoiBRaCB2RQguhI1Jo0Wx2HriqdQTdk4Niolk880ohJ87W4HojkkkDnWPQwRlJoXXozJkzJCcnYzAYtI4CQOn1/pS5JlJpttLaow37TnfWOpLTysvLY/78+Vy9ehUPDw9effVVpk+fbluXQuvQypUr6dGjB1OnTtU6CgAlF2D6BzWAB64udTze8bLWkZyS2WwmMTGRdevWMWLECEpKSggPD6dXr14MHjwYkELrVkhIiEMNQYRHmnl32Vk61G7B36ta6zhO49bRzjNnzmA2mxkxYgQAQUFB9OrVi0OHDtkKLQfFRLOIDPVm7d/CMLYxax3FaTQc7ezcpRvBwcFkZ2cDcPz4cY4dO8bQoUNtj5FCC/G7/Px8XFxcHObHNySZSxU3qDRbuXSllv/5+gLLli1jxowZBAQEEBYWxptvvkmPHj1sz0EKLcTvHH2088/9rzNx4kS2bNnCxYsXOXfuHKtWrSIrK8v2HKTQQjio+tHOzPSuHPrfKHb/cwddu3Zl0KBBAAQGBpKQkMD27dttj5FCC+HAbh3t7N27N0ePHqWoqAi4eXbSH374gb59+9ruL0e5hXAS/fv3Z9GiRSQlJXH9+nUAxo4dy+zZs233kUIL4URSU1NJTU2967q85RYt3qLV5+kxdi97TnTSOkqTyQ4tml1+fj6ZmZlaxwDgWFl7vj7QF1O1ldNu3fH3qdE6UpNIoUWzuvXPQo5BYbFYADfU7/92ZlJo0ezeeecdrSPYGZTzLxZ9VUKo/wG6O/n3zOUztGjxXkkK5OjGKPqHnNc6SpNJoXVmbe5FrSOIR2Tz5s0YDAa7H3d3d9LT0233kbfcOmGtU/RJ2kfFNSu116OYHrtX60jiDzZ69Ghqav590M5isdCvXz9SUlJst0mhG+nAgQMkJCTQqlUrraMAUNlqBLUdp1FpqsXQypO9p+QkAnpw6/ikt6eb3dp7771HdHQ0Q4YMsd0mhW6kjRs3MnHiRObMmaN1FAAuXa1jaPoJXF0NeLha6R1UrnUk0UT145OXKm6w8MsSDq3rj5fh5qfksrIyPvnkEw4dOmT3GCl0E7Rt25aQkBCtYwAQAvy6oTMLvyqF39bg1eqG1pGcTlxcHHFxcQ7zN/KGV8Zcur6UOZOCAFiwYAGTJ08mMDDQ7iFyUExHgoyt+eC1P+FruK51FKcUGxur+cjkvcYnp0+4+U02i8XCqlWr7vgVUNmhhXBQDa+M6dn65v6bm5tLQEAAvXv3vu0xUmghHFj9+OStduzYQXh4+B3vL2+5hXAy586do1OnOw+SyA4thJP58ssv77omO7QQOiKFFkJHpNBC6IgUWggdkUILoSNSaKGJ3Nxc+vbti7+/P927d+fTTz/VOpIuyJ+tRLMrLy9n3LhxZGdnM2rUKPbv38+QIUN46qmnGDhwoNbxnJrs0KJZmKqsfLy6FHO1lbq6OlauXMmoUaMA6NevH08++SRHjhzROKXzkx1aPHJ3GgOcMGECAFarlc2bN3P27FmGDRumcVLnJzu0DtWPAGp99cS7XUVx6fpSANauXUurVq1IS0vjs88+Izg4WONXzvlJoXXIWcYAJ06ciMVi4ZtvvmHu3Lnk5ORo/Mo5Pym0eOQaXkXx9MmjbNiwAQA3NzcGDBjAmDFj2Lp1q8ZJnZ9DFvofBVd59pVCdh6o1DrKbeqznbvcRusoTuXWqyheu3aNF198kby8PADOnDnDtm3beOqpp7QNqQPuAJs2bdI6h82pMjcyv/DnisnKPwuv8d8p5wjuZNA6FmCfzeARwbTB+7WO5JSio6NZsmQJaWlplJeX4+fnR3Jyst3paEXjuCilVGJiotY5bM6aevPr1WFcr3XHz9udyr3jGfvsAK1jAfbZWrtbiY84zsbl/6l1LCFs3OHmGSwdScp/Heb/Ck3ERPqS9Y8yrePYqc/m6XKKPl1KtY4jhB0X5ThXDbNzsNhMRHdvrWPc0cFiMxu+/ACllMNdp0m0bA55UAxw2DKDY2cTLZvDFloI8fCk0ELoiBRaaGb58uX4+vqycOFCu9v379/PwIEDCQgIIDQ0lKysLI0SOh8ptNDEzJkzyc3Nve380jdu3GDcuHGkpaVx4cIF1q1bx6xZs/j11181SupcpNCiWdw6PgkwefJksrOz8fPzs7vfjh07cHFxITU1FRcXFyIiIkhMTCQ7O1uL2E5HCi0eufrxyb/8/RThE/dRVVNHVFTUHe975MgRwsLC7G7r0aMHRUVFzRHV6UmhdSg/P1/zkckHGZ+8E7PZjKenp91tXl5emM3mR/2y6YIUWofy8vLIyMjQfGzyfuOTd+Lj40NVVZXdbSaTCR8fn0f9sumCFFo8cg3HJ+uvongnPXv2vO0AWGFh4V0vzibsSaFFs7h1fPJeBg8ejIeHBx9//DF1dXX89NNPbN26lZSUlGZK6tyk0EITBoMBg8FAbm4u8+bNw2AwkJ6ejpubGxs2bGDNmjW0b9+eGTNmsGLFCp544gmtIzsFOUmg0ERNTc1d1yIiIvj555+bMY1+yA4thI5IoYXQESm0EDoihdaZnQeuah1BaEgOiunIM68UcuJsDa43Ipk0sEDrOEIDUuhGCgkJYcGCBeTn52sdBYDS6/0pc02k0myltUcb9p3urHUkp7Z8+XLmzJnDX//6V+bMmWO7fdeuXcydO5cjR47g6+vLzJkzeeONNzRMak8K3UhTpkxxqEu3lFyA6R/UAB64utTxeMfLWkdyWjNnzuTChQu3fTutsrKShIQE3n//fVJTUykoKGDIkCFER0fz9NNPa5TWnhS6CeLi4rSOYCc80sy7y87SoXYL/l7VWsdxGqYqKys2lZOaaMTb043JkycTFRVFfHy83f0sFgsffvgh06ZNA6BPnz7069ePw4cPO0yh5aCYjkSGerP2b2EY28hk0oN6mNHOgIAAW5kBzp8/z8GDB4mNjW2uuPclhRbNyplHO29VWlrKqFGjePvtt3nyyScf8av24KTQolk582hnvf379zNo0CDS09N5/fXXm+FVe3DyGVq0aPWjnUvXlzJ9Qqf7ToPt3buX8ePHs3LlSoYPH95MKR+cFFq0ePWjnfdTXV1NcnIyK1ascMgyg7zlFuI2dxvt3L59OydPnmT06NG2+xgMBmbOnKl1ZBvZoYVo4F6jnQ56KTgb2aFFs1i0+jw9xu5lz4n7H3QSjSc7tE7l5+eTmZmpdQwAjpW15+sDfTFVWznt1h1/n7vvgKJppNA6dOufhRyDwmKxAG6o3/8tHg0ptE452nWrB+X8i0VflRDqf4Du8j3zR0Y+Q4tm8UpSIEc3RtE/5LzWUXRNCi3EHdztypj3W9OavOUWooG7jU/eb80RyA4tWrwHvTLm/dYcgezQokWrH5+8VHGDhV+WcGhd/7uOTwL3XHMEskOLZhUXF0dmZqbmY5NNHZ90VLJDi2YVGxvrQH8fB3O1lS4jv8fP2/eBxycdmRRatGgPOz7p6KTQosV70PFJZ+CiHOn9jxAOwGAwAHDjxg1cXV1xc3Nj2rRpLFmy5J5rjkAKLYSOOPcHBiGEHSm0EDrimp+fT15entY5mkwvz0OIpnDNy8vTRRH08jyEaAp5yy2EjkihhdARKbQQOuJ++vRpTp8+7TAnlGusvLw8QkJCtI4hhKbcAQc7oVzj6eE5CNEU7iEhIQQHBzvcSeUeVmZmphRatHjyGVoIHZFCC6EjUmghdMQ9Li5OF5899fI8hGgKGZ8UQkdsb7k//fRT+vXrh4eHB2lpafd80MqVK+nZsydeXl4YjUZSU1Mxm8229SNHjjB8+HB8fX0JDAxk7ty5WK1W2/qCBQvo2rUr3t7exMTEsGvXLtuayWRiypQpdOjQgbZt25KYmEhZWZltfdu2bURERODl5UVUVBQFBQV/yAshhC6o361fv15t2rRJTZ48WaWmpqp7OXjwoCoqKlKXLl1Sx44dU8OGDVPz5s2zrUdFRampU6cqk8mkiouL1eOPP64WL16slFJqy5YtysfHR+3evVvV1NSod999VxmNRlVXV6eUUmrevHmqT58+qqSkRFVUVKgxY8aopKQkpZRSJ0+eVJ6enmrdunWqqqpKLVq0SHXu3FlVVVXdM68QLYVthx4/fjxjxozB39//vr8EwsPD6dmzJ+3atSM4OBij0UhxcbFt/dSpU7z00kt4e3vTrVs3nnnmGYqKimxrffv2JTo6mtatW5OWlkZ5eTmXLl2yrY8ePZrOnTvTtm1bUlJSbI/97rvv6N27N88//zyenp7MmjULq9XKtm3b/tBfckI4q0Yf5d62bRuPPfYYfn5+bN68mVmzZtnWXnvtNbKysjCbzZw4cYLc3FyeffZZ4OYvjsuXL7Nnzx4sFgvLli0jOjqagIAAAGbMmMH333/P+fPnuXLlCl999RXPPfccAHV1dXfMcvDgwcY+DSF0pdFn/Rw6dCi//PILx44dY/Xq1QQHB9vW4uPj+eKLL/Dx8QEgPT3dVsqOHTsyevRoBgwYgFIKo9HI9u3bbY+NjIykXbt2BAXdPAtjdHQ0WVlZAIwcOZK5c+eyadMmEhISWLp0KWVlZVRVVTX2aQihK43eoVu3bk1gYCCxsbGMHDmSl19+GYCamhri4+NJTk6mqqqKkpISCgsLbV8tXbhwITk5ORQVFXH9+nU++ugjRowYwcWLFwGYNm0arq6u/Pbbb1y7do2YmBjGjBkDQPfu3cnJySEjIwOj0cjOnTsZMmSIrfxCtHgNP1TPnj37vgfFGlq7dq3q2rWrUkqpU6dOKUBVVFTY1pcsWaJiYmKUUkpNmTJFzZ492+7xRqNRffvtt0oppYKDg9WGDRtsa4cPH1aAqq6uvu3/NZlMql27dmrXrl0PlVcIvbLt0LW1tdTU1GC1WrFardTU1FBbWwtASUkJb7zxhu3fixcvprCwkCtXrrB7924yMjIYPnw4AF26dCEwMJDFixdjsVgoLy9n9erVREZGAhATE8OWLVs4fvw4VquVNWvWUFFRQVhYmG191apVVFRUUF1dzWeffUZoaKjtfMg//vgjly9fpri4mKlTpxITE0NMTExz/f4TwrHVNzsjI0MBdj+vv/66UkqpvXv32u2S8+bNU0FBQapVq1aqS5cu6tVXX1WVlZW23xJ79uxRgwcPVr6+viogIEClpKSoy5cvK6WUslqtav78+SooKEh5eXmpXr16qZycHNtjy8rKVFJSkmrfvr3y8/NTsbGxqqCgwLYeHx+vPD09VZs2bdTUqVPt3gkI0dLJN8WE0BEZzhBCR6TQQuiIFFoIHZFCC6EjUmghdEQKLYSOSKGF0BEptBA6IoUWQkek0ELoiBRaCB2RQguhI1JoIXRECi2EjkihhdARKbQQOiKFFkJHpNBC6IgUWggdkUILoSNSaCF0RAothI5IoYXQESm0EDoihRZCR6TQQuiIFFoIHZFCC6Ej/w+iJfaALOHxsQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log.ancestry().standardize()\n",
    "log.ancestry().render(\"%%inline\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX8AAABSCAYAAAC8AdQ7AAAACXBIWXMAAA3XAAAN1wFCKJt4AAAIa0lEQVR4nO3dbWhUVx7H8V9mgjvNZCbWjY+4SVCpFZslltoWt2p8aF+0hoWou+JKVERTFwMFUSl1q6Gl9IWWVIxCUYnC+hAsbhdKCZqalKIxsYK6ImvRDWrc0GikiTpjYvzvizaXzmbiZNuu6cz5fmBezPnfe+ecvPjl3nPvnEkzMxMAwCm+we4AAODxI/wBwEGEPwA4iPAHgCS2e/duhUIhVVRUxLQ3NDRo2rRpGjZsmHJzc7Vly5aYOuEPAEmqrKxMR48eVX5+fkx7R0eHXn31Va1YsULt7e365JNPVF5eri+++MLbhvAHgCRx516Pth24obuRHklSSUmJDh48qHA4HLNdV1eXtm7dqhUrVkiSCgoK9Oyzz+rixYveNoQ/ACSBO/d69Ns/fqW/VP5L+Qu/0r3oQ02dOjXuttnZ2Vq+fLn3vqWlRefOndPMmTO9tvT/e4+BQVJfX6/CwsLB7gbw8xi+QOGn3lDH3R5Jafro4xt6409jE+5248YNzZs3Txs3btTTTz/ttRP+SFl1dXXatGmTNm/ePNhdAX6yu5Ee/eblYwoHQ8rO8qt0wZiE+5w5c0bz58/Xm2++qVWrVsXUCH8ASALBJ/y6fuxlffTxDZUuGKMnfvXoWfumpiYVFxerqqpKc+bM6VMn/AEgSWQEfAOa6olEIlq0aJH27NkTN/glbvgCQNIKBAIKBAI6evSo1q1bp0AgoNWrV6umpkZXrlxRUVGRt00gEFBZWZm3L2f+AJCkotFov7VEa3Zy5o+U8+GBFk38fZMaLye+IQa4Kl2SysvLB7sfwM/iUuuv9fezU3Qn0qNm/wQ9mdn/mRHgsnQp8eUBkDxMXV1dkvyy798D6CtdEs9BI6X8rvrf+nD/dT315FlNGNE+2N0BfpGY80fK+fMfRuuff5uq5/JaBrsrwC8W4Q8ASay/JZ0T1XjUEwCSVFlZmdra2vos6ZyoJnHmDwBJY6BLOieqSZz5A0BS6F3S+dbtblX89br+cfi5fpd0lvTImkT4I4UVFhaqsLCQ77EgNfzIJZ37Q/gjZc2cOZPvsCBl/JglnR+F8AeAJPC/LumcCOEPAElioEs6D0SacV0MAEkpEAhIkrq7u+Xz+eT3+7V8+XLt3LnzkTWJ8AcAJ/GcPwA4iPAHAAf56uvrVVdXN9j9+MlSZRwA8Dj46urqUiI0U2UcAPA4MO0DAA4i/AHAQYQ/ADgovbm5Wc3NzUm/+FVdXZ3y8vIGuxsAkBS8H3BPhe96pcIYAOBxSM/Ly1Nubm7S/4h7eXk54Q8AA8ScPwA4iPAHAAcR/gDgoPTCwsKUmCtPlXEAwOPAks4A4CBv2ufOnTtavHixgsGgxowZo8rKyn53qqys1JQpUxQMBpWbm6t33nkn5qw7LS2tz2vNmjWSpM2bN/ep+Xw+tbW1SfruDP6/688884yk757lj3fsw4cPe5/9/vvvKycnR8FgUC+88IIaGhp+3r8YAKQC+97rr79uM2bMsG+++ca+/PJLC4VCdvz4cYvnrbfespqaGmttbbX6+nobPny47d6926tHIhHvdfv2bQuHw/b555+bmVl3d3dMvby83GbMmOHte//+/Zj6nDlz7O233zYzs56enpja8ePHLRgMWmdnp5mZffrpp5aZmWmnTp2yaDRq7777ro0cOdIePnwYdxwA4CqZmT148MBCoZDV1tZ6hdLSUlu6dOmADrJ69WpbuXJl3Nr+/fstNze33wCeOHGi7dmzJ27t+vXr5vf77fLly3Hrq1atspKSEu/99u3bbfr06d771tZWk2RtbW0DGgcAuMInSVevXlVnZ6cKCgq8K4KCggJduHBhIFcOamhoUH5+ftx6VVWVSkpKlJaW1qd28uRJXbt2TQsWLIi77759+zRt2jSNGzeuTy0ajerQoUNaunSp11ZcXKz29nY1Njaqq6tLu3bt0vPPP6/s7OyE4wAAl6RL3833S1I4HPYKWVlZ6uzsTHiATZs2ycy0cuXKPrWWlhbV1tZqx44dcfetqqrS/PnzFQqF4tb37t2r9evXx60dOXJEQ4cO1axZs7y2ESNGqKioSC+++KLMTCNHjlRNTU3CMQCAa3ySlJmZKUnq6OjwCt9++22/odxr69atOnTokD777DPvl+J/qPfMffz48X1q0WhU1dXVWrZsWdxj914VLFy4MG493hVFRUWFqqurdeHCBd2/f18ffPCB5s6dq5s3bz5yHADgGp8k5eTkKBQK6dy5c17h7Nmzmjx5cr87btu2TZWVlaqtrdWoUaPibrN3796YaZkfOnLkiMLhcMyZ+w9VVVWpuLg47j+g3iuKkpKSmPbz58+rqKhIkyZN0pAhQ7R48WL5/X6dPn2633EAgJN6J/9LS0tt9uzZduvWLWtsbLSsrCzvaZ9r167Z2rVrrbu728zMdu7caaNHj7aLFy96T950dXXF3Ew4ceKEZWRkWEdHR9ybDa+88opt3Lgxbi0SidjQoUPt2LFjcevvvfeevfTSS33ad+zYYRMmTLBLly7ZgwcP7ODBgzZkyBBrbm5OcOsDANzihX9nZ6ctWrTIMjIybNSoUbZ9+3Zvo6amJpNkkUjEzMzGjx9vkmJer732WsyBS0tLbcmSJXE/tPcpnq+//jpu/cCBA5aTk/PIJ4R27drVp72np8c2bNhgY8eOtYyMDJs8ebJVV1cn+BMAgHv4hi8AOIiF3QDAQYQ/ADiI8AcABxH+AOAgwh8AHET4A4CDCH8AcBDhDwAOIvwBwEGEPwA4iPAHAAcR/gDgIMIfABxE+AOAgwh/AHAQ4Q8ADiL8AcBBhD8AOIjwBwAHEf4A4CDCHwAcRPgDgIMIfwBwEOEPAA4i/AHAQYQ/ADiI8AcABxH+AOCg/wBkvS+uX/ErDwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log.ancestry(prune_death=True).render('%%inline')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATwAAABwCAYAAACD6MylAAAACXBIWXMAAA3XAAAN1wFCKJt4AAAaD0lEQVR4nO3de1SU5b4H8O/AyG1gBsQElA2muVNMhbyUhem2rNSz3ZoXMnVr3k3dW3fWcReaF2plppWmgkJuLRMsryXZERU7al6CABVS07w2KoIwXIbLDN/zB6s5TTPAyIwXmN9nLf6YeZ7n9zzv61o/38u8v1dBkhBCCCfgcq8XIIQQd4skPCGE05CEJ4RwGpLwRIOXmZmJwYMHw9/fH56enmjXrh1iYmJgMBjM+ul0OgQHB+Opp56qMVZKSgr8/Pwwb94803fNmjWDUqmEUqmEQqGAq6ur6fOaNWtM/SIjIxEcHAydTmc1dn3n/yN759m6dSuefPJJeHt7Q61W4/HHH0dSUlKN8zUmkvBEg5aWloYnnngCbdq0QU5ODgoLCxEfH4+kpCSMGTPGrO+GDRswYsQIFBQUICcnx2q82NhYLF26FOvXr4fRaAQA3Lx5EwaDAQaDAaGhodi4caPp86RJkwAAJ0+eRHl5OV544QVs3LjRauz6zv979s6zatUqjB07FuPGjcP169dx9epVzJw5E5MnT8batWutxmtUKEQDs3arltnnS0iSffv25YgRIyz6nD9/np9++imrqqpM33Xs2JEZGRlctmwZZ86caTFGq9WyZcuWrKio4MCBA7l9+3aLPqGhody0aZPF99OnT+dHH33E9PR0hoeHW123I+a3Z56SkhJqNBrGxcVZjNm/fz8PHDhgNV5jIglPNCj9pp+g+smDbPrUYR7NKqCbmxu//vrrOscdPHiQ3bt3J0nm5uYyKCiIZWVlZn1iYmL4+uuvkyR37NjB/v37W8SxlvBKSkoYEBDAvLw8kmRERASPHTvm8Pntnee7774jABYVFdW4nxo7OaUVd83GjRuhUCjq/+cVhszsPOhKjMjXVWJR7E+oqKhAixYt6pw7Li4OEyZMAFB9TS4yMhJffvmlqb2qqgrx8fEYP348AGDAgAHIysrCpUuX6oydmJiIp59+Gk2bNgUATJgwwezanqPmt3ee/Px8qNVqeHt717lNjda9zrjCeUyZMoWrV6+2K8afn99CdeQhNn3qEI+fKKC7uzu3bdtW65i8vDx6eXlRrVZTo9FQo9FQpVKxZ8+epj67du2iq6urqV2j0dDd3Z1z5841i2XtCK9bt25UqVSmcWq1mj4+PtTpdA6d3955Dh48SAC8detWPfZ84yBHeKJBOf3NC1j6r9b4fkMEuj6iQZ8+faxebL969SrGjRuHiooKrF+/HoMGDUJhYSEKCgpQUFAAnU6Hs2fP4qeffgJQfWS0evVqU3tBQQHS09PxySefWL158Jsff/wRWq0WRUVFpnGFhYV47rnnTDcVHDG/I+bp0qUL/Pz8rO6vb775Bu+9957d/z73vXudcYXzcMQR3h9lZGRQpVJx3LhxvHTpEvV6PQ8cOMD27dtz2rRpJMl27dpxx44dFmNfeeUVzpo1i5cvX6aHhwdv3rxp0ScsLMxs7B+P8CZNmsRZs2ZZjNu8eTMjIiIcNr8j5iHJ+Ph4uru7c+nSpczPz6dOp+Nnn31GX19fJiYmWoxtbCThibvmTiQ8kszKyuLf/vY3+vn50cPDg4888ghXrlzJqqoq7t+/nxqNxuIGAUmmpqbS39+f0dHRfP75563Gnj9/PgcMGGD6/PuEp9Pp6O3tzSNHjliMKykpoUqlYmxsrN3zP/fccw6Z57e27du3s0ePHvTy8qJGo2GvXr24Z88eq/M3NgpSigc0FpUGoolS4bB4JKBwUDhjFTF92ivo3LkzpkyZ4pigQtwmuYbXSPzP97fQadgPmPr2WbtjlVcQ7Qb9gD//7TjydYa6B9ThzZW/4JEhaTiXF2p3LCHsobzXCxDA9evXERsbW+/xBaXuiDvwJPTlxLkrZbiSvRld/1z/RLXhSG+cv1Z9aNfu2XWY1u/Xesc6ee1B7PqxNfTlVTjn0g1dO12pdywh7CVHePeBsWPHIicnB6y+pnrbfwBQWVmd4EgFvv5qZ71jkQR+fxrLKsyfP7/+sUhU/vZMq0JRfZ4sxD0iR3j3AQ8PD7z44osYNGhQvWP817ECzHj3LPp098PHc47atZ43DUTn4WkwGoljn4+HRjXZrnhvrbqAzSk3EdJkD0Ie8LMrlhD2kITXSDzd3ReZm7s65KZFE6UC2Vu7OuymxYJXWmHu5FDMmJ4AQBKeuHfklLYRceQdWsBxd2gBQOnq2LU5K1vKR0VHR8PPzw/79u2z2m6tfNTy5ctNJa9cXV2hUChMn4ODg+tst2V9BoMB7733Hh555BF4enrCz88PPXr0wKZNm2zajszMTNOcv5Xq+v3n06dP17h9Jnfkxy7itgwaNKjOx6Magzv1OzxnMmTIECYkJDAkJIQGg8GivbKykqGhoVy7di2HDx9uNcaKFSs4e/ZsduzYkdnZ2Rbt69at48MPP1zjGmprr219w4YN48MPP8zU1FRWVlayqKiIX3zxBZs1a8Y1a9bc1nb88ssvBECtVntb2ydHeKLBid92DTm/lDok1uFMHTYm33BIrOJSI5Zv+hUl+pofRbsdn359HUdPFJk+X7t2DUeOHMHo0aMRHh6Or7/+2mLMjh078Oijj+Lvf/87Dh48iNzcXIs+a9aswahRo/Dyyy9bFB+wR23rO3ToELZt24avvvoKvXr1glKphLe3N4YOHYrMzEyMHj36trejJrVtnyQ80aD0n3ESry49h8ixmfjhVLFdsXYfvoV+005g6ttnMTnmjF2xikuN6BSVhrkrf0HHYWkoLauyK96EBWcw7d1zeHZqFv7nSD4AICEhASNHjkSTJk0wfvx4q8kqNjYWEydOhJubG6KiovCf//zHrP3QoUPw9PRE586dMXr0aCQlJaG8vNyutf6mtvXt27cPXbp0Qdu2bS3GtWjRAh4eHre1HTWpa/sk4Ym7JjIyElOnTnVYeagB/Z+1q9zU2NdToCsxoqjUiB1bP7crlk+rF5F3qxK6EiPyCgxQebrWP557C+w6cBlFJQboSoyIWXvZpvJR586dw5kzZ/Dcc88BqC4ftXbtWtNPl4C6y1TVV13ry8/PR2BgoNmYRx99FIGBgQgMDESnTp1uaztqUtf2ScITd83IkSPt+n0gS7Ph7ZYHtbcSTdVK3MjNsyvep+8/A3dXI3y8XTFk2CgAqHes4ouJcEUx1ColmmlcARf3+q+t/Fe0VOdCrXKFWuWKtyaFYPfu3bh8+TK6d+8OX19f+Pv7Izc3F/Hx8ab9GxcXh+vXr6Np06bw9fXFE088gfPnz2P//v0AqpPOli1bMHv2bPj6+sLX1xfJycmIi4uz+9+2rvU1b94cly9fNhvz7bffIiMjA+vXr8eNG/9/WaGu7aiJTdtnccVP3HXOctPCUdZu1fL0hVKHxDqSpeNnu647JFaJ3sgPPrvM0jKjQ+J9tus6j52srnU3cOBAiwv7p06dYsuWLWkwGFheXs4HHniAZ86cMevz8ccfMyoqiiS5bNkyvvTSS2btRqORgYGBzMnJMX1Xn5sWda0vLS2NLi4uzMjIsIh37NgxBgQEkKRN20Fav2lhy/ZJwrsPSMITtbGlfNTnn39uKhP1e9euXaOnpydzc3NtKh9F3n7Cs7W81vjx4xkSEsLk5GSWl5ezpKSEycnJ7NSpE/v160eSNm0HaT3h2bJ9ckorxH0uPj4evXv3hr+/v0Xb8OHDsWbNGsTFxSEqKsqiPSAgAI899hjmzp0LrVZrui72xxgbNmyo980LW9YHVN89nT17NubMmQONRoPAwEAsXLgQU6dOxc6dOwGgzu2o6eZFamqqTdsn5aHuA4MHD8aYMWPserRMCFE3OcITQjgNSXhCCKchCU8I4TQclvDOXSlzVCgAwMlzjnl0CAC+z9I5LNbNgkpUGhx32fOHbPueFhBC2M4h5aHWf30Nry79BfoyI05v74bgAPd6x8q9VYlHhv4AVxfgX6P/hNl/D657UC1G/DsHx07o8EBTN/zvJ+F2VRTZc6QAg/51CmXlRmRv6YqHW3nVO1ZRiRFdXkpHWUUV3PW96h1HCGE7BUn+5S9/qXeASnrhJ76OPF11Immt/BIh3ifqHU/rPgWnrwcBAB5wP4cHFevg5eFar1i5FR1ww20EcguMcHdzQZvKRWiuqahXrCo0QQ7+jdyC6rUEK3ejjer7epdQuuH2En7KbYcqEh6uxZg7mnjjnwPqF6wW0dHRWLlyJbZs2YI+ffpY7RMZGYkLFy4gOzsbarXaoj0lJQXDhg3DjBkzsHDhQgDAkCFDsGPHDgDVjxUBgItL9QlDVFSU6V2ptc1vMBiwbNkybNiwAefOnYOHhwfatWuHf/zjHxgxYsRtr1On0yEsLAytW7fGd999Z9G+detWLF26FJmZmXBxcUFYWBhmzZpl8TOIuuKIhktBkqmpqXYFefGtItwo1kDpokDlL0uw/8t/1TvWgZPeiFlXAkOVAqEBFbj4bd86HympSUExMHKRAaUVbtD4uKLw+Ejs/+Y/9V7buJhcXMwPhEIBGH/9BPs3jal3rB/Pe2LO6lJUGFzh5qpHUrTe4T9LMRgMeOihhxAdHY09e/YgKSnJos/Jkyfx8ssvo0ePHmjfvj2mTp1q0Wfo0KHo378/FixYgPPnz8PV1fw/oLFjx8LDw8PivRx1zT98+HBkZWUhLi4OTz75JMrKyrB7925MnToV77zzDiZOnHhb6/z4449x8eJFfPvtt0hKSkL79u1NbatWrcKcOXPwwQcf4MUXX0RVVRV27dqFKVOmYMmSJWZz1RZHNHA1/pz6NujLjRw37ye++8klR4Tj2q1ajn4zh4XFlvW+btfZS6Uc/no2/ze9wO5YBmMVp8Sc4exl5+yORZLb9ubyxf/OYb+/vmT3kxZVVeTHSVd5Sfv/7yX98ssvOXjwYJaXl7NFixa8ceOGxbjp06fzo48+Ynp6OsPDwy3atVotW7ZsyYqKCg4cOJDbt2+36DNmzBhOnjzZ4vva5j948CCVSqXFI0QkefXqVer1+ttaJ0l27NiRGRkZXLZsGWfOnGn6vqSkhBqNhnFxcRZj9u/fzwMHDtgURzR88mjZfcARj5Z1G5lOTeRBNn/6MM9crH7O9JlnnmFycjJJctasWXzvvffMxpSUlDAgIIB5eXkkyYiICB47dsysT0xMDF9//XWS5I4dO9i/f3+LuWtKeLXNv3DhQj722GM2bZst6zx48CC7d+9OkszNzWVQUJDpxdPfffcdAbCoqKjOuWqLIxo++VnKfaCwsBCDBw+ufzkh3yfw6/VSFBYbcSOvEu9vuGJTiZ3ExEQ8/fTTaNq0qanP72uY2VKSqCZ1zX875YLqWidQe1mg/Px8qNVqeHt717nuO1U+Sdwn7nXGFY45wmv5l6+ofvIQm/c5zJ8vl/K1116ju7s7NRqN6c/V1ZV79+41jenWrRtVKpWpXa1W08fHhzpddYWOXbt20dXV1SyGu7s7586daza3tSO8uuZ/5513+Oijj5qNuXHjBrVaLXfv3m2qnmHLOvPy8ujl5UW1Wm3qo1Kp2LNnT5LVR20AeOvWrVr3YV1xRMMnCe8+4KhqKR8nXeWV62U2ldhJT09ncHAwq6qqzPoMHTrU9N6Jukr+/OaPCc+W+W0tF2TLOusqC6TX6+nn52dxSk+SycnJXLx4sU1xRMMnCe8+4OjyULaU2Jk0aZJZSaDfbN68mRERETaX/CEtE56tJX5sKRdU1zpJ28oCxcfH093dnUuXLmV+fj51Oh0/++wz+vr6MjEx0eY4omGThHcfcHTC69WrF999912rbb1792ZMTAy9vb155MgRi/aSkhKqVCpOnDiRzz//vNUY8+fP54ABA0yf/5jw6pp/yZIlJKuPnpYvX85OnTrRw8ODPj4+fPzxx7l69WpWVlZSp9PVuc7Y2FhqNBqrNxZSU1Pp7+9vatu+fTt79OhBLy8vajQa9urVi3v27CFZfbfW1jii4ZLyUPcBKQ8lxN0hd2mFEE5DEp4QwmlIwhNCOA1JeEIIpyEJTwjhNCThNXLR0dHw8/PDvn37auwTGRmJ4OBg6HTWC6WmpKTAz88P8+bNM303ZMgQKJVKKJVKuLi4wMXFxfR55MiRd3z+29lOg8GAxYsXIywsDB4eHmjWrBn++te/Ii0tzazftm3b0LNnT/j4+EClUiEsLAyLFi1CZWVljfOKBuZe/y5G3Ln30lZWVjI0NJRr167l8OHDrfY5ceIEu3btyhkzZnDVqlVW+wwZMoQJCQkMCQkxe8LiNzUVD7hb89c1z7Bhw/jQQw9x3759rKys5LVr1zh//nx6enoyKyuLZPVTIBqNhuvWrWNxcTHLysr4/fffs0uXLhZPX4iGSxLefcARCa+hlYdy5Py1zXPo0CG6urry1KlTFuM2btzIM2fOsKSkhGq1muvWrbPoc+vWLVOVFtHwySltI/HY6B/x5opf0HVUOs5e0gMAYmNjMXHiRLi5uSEqKsriJcalpaX44osvMGrUKEREREChUOD48eNmfRISEjBy5Eg0adIE48ePt6hSUpu7NX9t8+zduxcREREICwuzGPfSSy+hbdu2SEtLQ1FRkdUXQPv6+pqqtIhG4F5nXEEuXryYAOr/p+nBls8cJMIPEOEHOGnRGf78888MCQmh0WgkWf3Qf9u2bc0ewk9ISDA7XVu5ciUnTJhg+mw0GtmqVSuePn2aJGkwGBgcHMyLFy+ard/aEd7dmr+ueWbOnMmBAwfWuv+3b99OjUZj9t20adMYEBBg+jt79mytMUTDIAmvkWho5aEcNX9d88TExLBz58617rvDhw8TAAsLC03fFRQUUKvVUqvV0svLS6qlNBKS8BqRhlQeyhHz2zLP0aNHqVAoePToUYv99e677zI5OZllZWVs1qwZly9fbtGHJFUqlSS8RkISXiPUEMpDOWJ+W8tQjRo1ikFBQdy5cyf1ej21Wi2jo6Pp7+9vOl3esGEDvby8uGLFChYWFrKiooJpaWl84YUXGBwcbHUdouGRhNcI3e/loRw1v61lqAwGAxcvXsz27dvT3d2dzZs3Z1RUlMV1ud27d7N379708fGhp6cn27dvz3//+99yl7YRkfJQQginIT9LEUI4DUl4Qgin4XLgwAGkpqbe63XYrbFshxDiznFJTU1tFImisWyHEOLOkVNaIYTTkIQnhHAakvCEEE5DeeHCBVy4cAELFiy412uxS2pqKlq1anWvlyGEuI8pAYDVT1zc67XYrTFsgxDizlG2atUKoaGhmD9//r1ei10WLFggCU8IUSu5hieEcBqS8IQQTkMSnhDCaSh79+7dKK59NZbtEELcOVIeSgjhNEyntP/85z8RGhoKDw8PPPjgg1iyZEmtAxctWoTmzZtDrVZjypQpZi8rHjRoEBQKhenP29vb1JaRkWHWplAo8P7775vaP/zwQ4v2H374wdRuNBrxxhtvICgoCJ6enujevTuKi4sdsjOEEI2bKeENGzYMqamp0Gq1SExMxIcffojk5GSrg5KSkrBy5UqkpKQgOzsbhw8fxttvv23WZ/Xq1dDr9dDr9cjLyzNra9mypalNr9dj1qxZZu0jRowwa+/SpYup7c0338SePXuwc+dOXLx4Ee+88w6USqXdO0II0fiZMkVkZKTpyzZt2sDb2xs///yz1UHr16/HpEmT0KlTJwDAG2+8gTlz5pj9lk+pVMLDw6PGiWtrc3Fxsdqel5eHFStWICMjA23btgUAPPPMMzXGEUKI3zO7Sztv3jw0b94cgYGBqKqqsvpiYgA4deoUwsPDTZ/Dw8Nx8eJFs1PLBQsW4E9/+hP69OmDvXv3mo2/efMm2rRpgw4dOuDVV19FUVGRWfvu3bsRHByMbt26YeXKlaabEZmZmfD390d8fDz8/Pzw0EMPYdWqVfbtASGE0zA7F3zttdcwceJEHD58GEePHoVarbY6qLi42KxNo9GYvvf29sbkyZPRvHlzqFQqfPXVV+jfvz+OHj2K8PBwBAYGIjExER06dMCvv/6K2bNnY/z48di8eTMAoGfPntiyZQtatmyJH3/8EdOmTYNCocArr7yCK1eu4MqVKyguLsaVK1dw4sQJPPvss2jbti369u17p/aREKKxqOntPtOnT+f8+fOttoWEhHDLli2mzzk5OQTAoqIiq/0HDhzI6Ohoq23p6elUKpUsKyuz2r5s2TJGRkaSrH6FHwCzV+aNHTuWM2fOrGkzhBDCpMYfHpPE+fPnrbZ16NABWVlZps+ZmZkIDQ01uxv7e25ubjAajTW2VVVVoaqqqs6xv10zFEKIeiHJ4uJifvjhhzx37hzz8vK4detWqlQqrl+/niR5+fJlvvrqq6ysrCRJbtq0iUFBQczOzqZWq2VERATfeustkmRpaSk//fRTXr16lTdv3uQnn3xCNzc3Hjp0iCSZkpLCtLQ03rp1iydPnmTv3r3N3j+amJjIs2fPsqCggPv27WNwcLDZu0d79OjBGTNmUK/X8/jx41Sr1UxJSbkb/zkIIRo4kKRer2e/fv3YtGlTenh4sF27dvzggw9MnY4fP04A1Ov1pu8WLFhAf39/+vj4cNKkSayoqCBZnfCeeuopajQaenp6Mjw8nF988YVpXEJCAlu3bk03NzcGBQVx7NixvHHjhql92rRpDAgIoJubG1u3bs2FCxfSYDCY2i9dusS+ffvS09OTDz74IGNjY+/c3hFCNCrypIUQwmlI8QAhhNOQhCeEcBqS8IQQTkMSnhDCaUjCE0I4DUl4QginIQlPCOE0JOEJIZyGJDwhhNOQhCeEcBqS8IQQTkMSnhDCaUjCE0I4DUl4QginIQlPCOE0JOEJIZyGJDwhhNOQhCeEcBqS8IQQTkMSnhDCaUjCE0I4DUl4QginIQlPCOE0JOEJIZyGJDwhhNOQhCeEcBqS8IQQTkMSnhDCafwfKLvuDpOBexcAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log.mutations().render(\"%%inline\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAASQAAABwCAYAAAC3kQ25AAAACXBIWXMAAA3XAAAN1wFCKJt4AAAXkUlEQVR4nO3de1RU5d4H8O8AAsNlJoQEhAOm+WooCpqmiZdjmakr85KyvB0t8JLWOvimrU5piXo8Xorj0VRU0DRNUFOxJFfewBc9XkIBU09eymuYgnKVy8zwff/wdb9OM8DIxRno91lr/tjzXH7Ps13+1t4Pez+jIkkIIYQNsLP2AIQQ4iFJSEIImyEJSQhhMyQhNRKzZs2Ch4cHDh48WGmdsLAw+Pv7o6CgwGz5/v374eHhgY8//lj5bvjw4XBwcICDgwPs7OxgZ2enHI8ZM8ai+Hq9HosXL0b79u2hVqvh4eGB7t27Y8uWLTUaZ0FBAfz9/dGrVy+z5Tt27ECPHj3g5uYGjUaDbt26ITEx8bH7EVZA0eDpdDoGBgZy7dq1HDlypNk6Z86c4fPPP893332XK1euNFtn+PDhjI+PZ0BAAPV6vUn5+PHjOXny5MeOP2LECLZp04YpKSnU6XQsLCzktm3b6OXlxTVr1jz2OJcvX84ZM2YwODiY586dMypbsWIF3d3dGRcXx6KiIhYUFHDLli3UarUmsarqR1iHJKQGpqKC/DzxJq9llyrfbd++nUOHDmVZWRmbN2/O27dvm7R75513+K9//YunTp1iSEiISXl2djb9/PxYXl7OwYMHc9euXSZ1KktIVcVPS0ujg4MDL1y4YNLu5s2bLCkpeaxxkmRwcDAzMjIYExPDqKgo5fvi4mJqtVquXr3apM2hQ4eYmppqUT/CeiQhNTBdxpyiNiyNzV46ygtX75MkX375ZSYnJ5Mkp0+fzsWLFxu1KS4upre3N3Nzc0mSoaGhPHHihFGd+fPn8/333ydJJiUlceDAgSaxK0tIVcWfO3cuX3jhBYvmZsk409LS2LVrV5LknTt36Ovry9LSB8n58OHDBMDCwsJqY1XVj7AeWUN6ghYvXgyVSlXzz1Mv4tff7iO/yIDbuTp8uvEGLl++jAsXLqB///4AgMjISKxduxZ85PGyhIQEvPTSS2jatKlSZ82aNUp5RUUF4uLiEBERAQAYNGgQsrKycO3atWrnVF38u3fvwsfHx6hNp06d4OPjAx8fH3To0MHicQLA6tWrERkZCQDw8vJCWFgYtm/frsTSaDRwc3OrdtxV9SOsyNoZ8Y9kyJAh3LlzZ6368PvzN9T0OMJmfY/y0vX7nDlzJp2cnKjVapWPvb09Dxw4oLTp0qULXV1dlXKNRkN3d3cWFBSQJPfs2UN7e3ujPpycnDh79myj2OaukKqLv2DBAnbq1Mmoze3bt5mdnc29e/fS29vb4nHm5ubSxcWFGo1GqePq6sqePXuSfHDVA4D37t2r8hxW14+wHklIT1BdJCTywRrSjd9KWVZWxqefftpkfebzzz9neHg4SfLUqVP09/dnRUWFUZ033niDq1atIkkOHjzYZMH37Nmz9PPzM1rc/n1CsiR+eno67ezsmJGRYTKPEydOKAnJknHGxMRw9OjRRuUGg4E+Pj48f/48S0pK6OHhYXLLSpLJyclctGiRRf0I65GE9ATVVUJ66KuvvmJoaKjJ97du3aJareadO3c4adIkTp8+3aTO1q1bGRoayuvXr9PZ2Zk5OTkmdYKCgpiUlKQc/z4hWRKfJCMiIhgQEMDk5GSWlZWxuLiYycnJ7NChAwcMGECS1Y6TJNu2bWs0noemTp2qtI2Li6OTkxM/++wz3r17lwUFBdy0aROfeuopJiQkWNyPsA5JSE9QXSek3r17c+HChWbL+vTpw/nz59PNzY3Hjh0zKS8uLqarqysnTpzIV1991Wwfc+bM4aBBg5Tj3yek6uIvWbKE5IOrj2XLlrFDhw50dnamu7s7u3XrxlWrVlGn07GgoKDaccbGxlKr1ZpdeE5JSaGnp6dStmvXLnbv3p0uLi7UarXs3bs39+3bR/LBX9ss7Uc8eSpSXq59EnR6YuSIYRg/fjyGDBli7eEIYZMcrD2Axo4EJs2/iP85nQ+v0gBrD0cImyYJqZ7Nib2CDbt/g05fgUuqIcit5HUIIYS8y1bvtm5NhE6nAwAQhEqlsvKIhLBdcoVUz9o+dQyBrTvhWo47PEs3oql7X2sPSQibJVdIT8CUgcXISOwML6fr1h6KEDZNEtIT4tikfm/VbHn7kdrEf5x56vV6LFq0CEFBQXB2doaXlxdee+01pKenG9XbuXMnevbsCXd3d7i6uiIoKAjz5s1Tbq2FFVn7uYPG7tFnj+r6OaSHbH37kbqKb8k2J88++ywPHjxInU7HW7ducc6cOVSr1czKyiL54ClyrVbL9evXs6ioiKWlpfz3v//Nzp07mzy9LZ48SUj1rK4TUkPbfqQu41cV58iRI7S3t+fZs2dN2m3evJkXLlxgcXExNRoN169fb1Ln3r17yi4Dwnrklq2BeWHcaXy0/Bc8P/YULl4rAQDExsZi4sSJcHR0RHh4OL744gujNvfv38e2bdswduxYhIaGQqVS4eTJk0Z14uPjMWbMGDRp0gQREREmb9lX5UnFryrOgQMHEBoaiqCgIJN2o0ePRuvWrZGeno7CwkKEh4eb1HnqqaeUXQaEFVk7IzZ2j14VLVq0iABq/tF2p9/LaURIKhGSyknzLvDSpUsMCAigwWAg+eCl2NatWxu9pBofH290O7JixQpGRkYqxwaDgS1atOBPP/1EktTr9fT39+fVq1eN5mLuCulJxa8uTlRUFAcPHlzlv8WuXbuo1WqNvps2bRq9vb2Vz8WLF6vsQ9QvSUj1rK7XjRra9iN1Fb+6OPPnz2fHjh2rPHdHjx4lAObn5yvf5eXlMTs7m9nZ2XRxcZG3/a1MElI9q4+F7Ia0/UhdxLckzvHjx6lSqXj8+HGT87Vw4UImJyeztLSUXl5eXLZsmUkdknR1dZWEZGWSkOpZff1ljWwY24/URXxLtzkZO3YsfX19uXv3bpaUlDA7O5uzZs2ip6encju4ceNGuri4cPny5czPz2d5eTnT09M5bNgw+vv7mx2HeHIkIdWz+kxItr79SF3Ft3SbE71ez0WLFvG5556jk5MTmzVrxvDwcJN1ob1797JPnz50d3enWq3mc889x7/97W/yVzYbUGfbj+j0RBMHeU/rUbLliBCPp07+7H/5Rinav5GOsDczYKiQ7ZVIYOK8iwgemY47suWIEBZzAIDo6OhadbJgT1+U64gLV4Gury3D4K55dTK4hurk9WB8n9lMthwR4jE5AEBt79oMBj0AewDAqfQf8FqXVrUeWMNWAYPBAEAlW44I8RjqZA3pyq9leHXaGXg3bYJDazvCTp7/xtsLLiH1hzx4lm7Ee5P6yhqSEBaQRe16VK4jwkfKorYQlqqzaxlJRqbqe8uRxiQzMxNDhw6Fp6cn1Go12rZti/nz50Ov1xvVKygogL+/P3r16lVpX+a2MfHy8lK2TVGpVLC3t1eOH31vrrotUmoa//dqG2fHjh3o0aMH3NzcoNFo0K1bNyQmJlYar6GQmythdenp6XjxxRfRqlUrnD9/Hvn5+YiLi0NiYiLGjx9vVHfjxo0YNWoU8vLycP78ebP9xcbG4rPPPsOGDRv+by0PyMnJgV6vh16vR2BgIDZv3qwcT5o0CQDw448/oqysDMOGDcPmzZvN9l3T+I+qbZyVK1diwoQJeOutt/Dbb7/h5s2biIqKwuTJk7F27Vqz/TUYVnwG6g+hPh+MbMjW7sjmuZ+LSZL9+vXjqFGjTOr8/PPP/PLLL41eOwkODmZGRgZjYmIYFRVl0saSbUwCAwO5ZcsWk++r2yKlruLXJk5xcTG1Wi1Xr15t0ubQoUNMTU01219DIQmpnklCMjXgnTPU9Ehj015HeTwrj46Ojvz222+rbZeWlsauXbuSJO/cuUNfX1+TH3WcP38+33//fZJkUlISBw4caNKPuYRUXFxMb29v5Wnt0NBQnjhxos7j1zbO4cOHCYCFhYWVnqeGTG7ZxGPZvHkzVCpVzT8uQcg8l4uCYgPuFugwL/Y/KC8vR/PmzauNvXr1akRGRgJ4sCYUFhaG7du3K+UVFRWIi4tDREQEAGDQoEHIysrCtWvXqu07ISEBL730krInUmRkpMmeTHURv7Zx7t69C41GAzc3t2rn1CBZOyM2do3tCmnKlCnKW/o19V+vfk1N2BE27XWEJ8/k0cnJqdpzlJubSxcXF2o0GmX7EVdXV/bs2VOpY+k2KuaukKrbIqWu4tc2TlpaGgHw3r17NTjztk+ukMQT99N3w/DZf7fEvzeG4vn2WvTt29fsYuzNmzfx1ltvoby8HBs2bMCQIUOQn5+PvLw85OXloaCgABcvXsR//vMfAA+uLFatWqWU5+Xl4dSpU1i3bp3ZxeWHTp8+jezsbBQWFirt8vPz0b9/f2XRuS7i10Wczp07w8PDw+z5+u6777B48eJa//tYkyQkYRWRQ33wX4FqAMA//vEPpKamIiIiAtevX0dpaSkOHz6Mfv36wcXFBY6OjlizZo3J1rN2dnYYNmwY1qxZgxs3buD777/HsGHDjOoEBQVBq9Viz549lY4lNjYWI0aMMHmifuTIkcrtVF3Er4s4zs7OWLJkCWbPno2YmBjcu3cPhYWF2Lx5M0aPHo3AwMBK59kgWPsSrbGTWzbLZGVl8fXXX6eHhwednZ3Zvn17rlixghUVFTx06BC1Wq3JAjJJpqSk0NPTk7NmzbJ4G5VHb9kKCgqq3SIlNja21vH79+9fJ3Eelu3atYvdu3eni4sLtVote/fuzX379pmN35DU2ZPawryhQ4c2mie1DRXEO9OmomPHjpgyZYq1hyMaIbllExb5aMUvaD88HZdzG/gtgbBpDtYeQGNXVlaGxMREZGZmWnsoNfbjrWew53RLlJRV4LJdFzzf4Ya1hyQaKblCqmfr1q1DmzZtwAcPoTbID0joHr5TplI92IFOiHoga0jCIp+svIKt+3MQ0GQfhv7ZQ9aQRL2QKyRhkeipLXBmW2e0bFr9U89C1JQkJGExB3vZTqU6lmw9MmvWLHh4eODgwYNmy81tPbJs2TJluxR7e3uoVCrl2N/fv9pyS8an1+uxePFitG/fHmq1Gh4eHujevTu2bNli0TwyMzOVmA+3eXn0+Keffqp0foo6foxANHL19RxSYzF8+HDGx8czICDA6Ec2H9LpdAwMDOTatWs5cuRIs30sX76cM2bMYHBwMM+dO2dSvn79erZp06bSMVRVXtX4RowYwTZt2jAlJYU6nY6FhYXctm0bvby8TH7Is7p5/PLLLwTA7Ozsx5qfXCEJUUNffvsbjp8pVI5v3bqFY8eOYdy4cQgJCcG3335r0iYpKQmdOnXCX/7yF6SlpeHOnTsmddasWYOxY8fizTffNHnxtjaqGt+RI0ewc+dOfPPNN+jduzccHBzg5uaGN954A5mZmRg3btxjz6MyVc1PEpIQNRAZfQHTFl7GK29n4ftjdwEA8fHxGDNmDJo0aYKIiAizySQ2NhYTJ06Eo6MjwsPD8cUXXxiVHzlyBGq1Gh07dsS4ceOQmJiIsrKyOhlzVeM7ePAgOnfujNatW5u0a968OZydnR9rHpWpdn6VXvcJYcamTZsI4I/9cfSlz5/3EyGpREgqe76VQYPBwBYtWig/2a3X6+nv78+rV68q5+7SpUsMCAigwWAgSZ49e5atW7c22oBu3LhxRrdHI0aM4KZNm4z+DWpyy1bd+KKiovj6668btQkNDaW3tze9vb0ZHBz8WPOo7JatuvlJQhKiBjq/voWaHmnU9Ejj/mN3Ldp6ZObMmXRycjKqY29vzwMHDpC0bIsTsmYJqbrxLViwgJ06dTJqc/v2bWZnZ3Pv3r309va2eB6k+YRkyfwkIQlRQ5v2/MYTPz7Yx2jw4MEmC79nz56ln58f9Xo9y8rK+PTTT/PChQtGdT7//HOGh4eTJGNiYjh69GijcoPBQB8fH54/f175riYJqbrxpaen087OjhkZGSb9nThxQklIlsyDNJ+QLJmfJCQhaun69et0dnZmTk6OSVlQUBCTkpL41VdfMTQ01KT81q1bVKvVvHPnDtu2bcukpCSTOlOnTuX06dOV48dNSJaMjyQjIiIYEBDA5ORklpWVsbi4mMnJyezQoQMHDBhAkhbNgzSfkCyZnyxqC1FLcXFx6NOnDzw9PU3KHu51tHr1apN9jgDA29sbL7zwAmbPno3s7Gz079/fbB8bN26s8eK2JeMDHvz1a8aMGfjggw+g1Wrh4+ODuXPn4u2338bu3bsBoNp5VLa4nZKSYtH85NURIYTNkCskIYTNkIQkhLAZdqmpqUhJSbH2OGqtscxDiD8yu5SUlEbxH7mxzEOIPzK5ZRNC2AxJSEIImyEJSQhhMxyuXLmCK1euIDo62tpjqZWUlBS0aNHC2sMQQtSCA4D/38y9gWsMcxDij8yhRYsWCAwMxJw5c6w9llqJjo6WhCREAydrSEIImyEJSQhhMyQhCSFshkOfPn0axdpLY5mHEH9ksv2IEMJmKLdsf/3rXxEYGAhnZ2c888wzWLJkSZUN582bh2bNmkGj0WDKlCnQ6XRK2ZAhQ6BSqZSPm5ubUpaRkWFUplKp8OmnnyrlS5cuNSn/4YcflHKDwYAPP/wQvr6+UKvV6Nq1K4qKiurkZAghrEtJSCNGjFB2dUtISMDSpUuRnJxstlFiYiJWrFiB/fv349y5czh69Cj+/ve/G9VZtWoVSkpKUFJSgtzcXKMyPz8/paykpATTp083Kh81apRReefOnZWyjz76CPv27cPu3btx9epVLFiwAA4ODrU+EUII61P+J4eFhSlftmrVCm5ubrh06ZLZRhs2bMCkSZPQoUMHAMCHH36IDz74wOhZJgcHB5PfcnpUVWV2dnZmy3Nzc7F8+XJkZGQovx/18ssvV9qPEKJhMfor28cff4xmzZrBx8cHFRUVZvfOBYCzZ88iJCREOQ4JCcHVq1eNbp2io6Pxpz/9CX379sWBAweM2ufk5KBVq1Zo164d3nvvPRQWFhqV7927F/7+/ujSpQtWrFihLFZnZmbC09MTcXFx8PDwwLPPPouVK1fW7gwIIWyG0b3OzJkzMXHiRBw9ehTHjx+HRqMx26ioqMioTKvVKt+7ublh8uTJaNasGVxdXfHNN99g4MCBOH78OEJCQuDj44OEhAS0a9cOv/76K2bMmIGIiAhs3boVANCzZ098/fXX8PPzw+nTpzFt2jSoVCpMnToVN27cwI0bN1BUVIQbN27gzJkzeOWVV9C6dWv069evvs6REOJJqeynVN555x3OmTPHbFlAQAC//vpr5fj8+fMEwMLCQrP1Bw8ezFmzZpktO3XqFB0cHFhaWmq2PCYmhmFhYSTJrVu3EoDRz7lMmDCBUVFRlU1DCNGAVPpgJEn8/PPPZsvatWuHrKws5TgzMxOBgYFGf017lKOjIwwGQ6VlFRUVqKioqLbtwzUrIUQjRZJFRUVcunQpL1++zNzcXO7YsYOurq7csGEDyQc/NPfee+9Rp9ORJLds2UJfX1+eO3eO2dnZDA0N5SeffEKSvH//Pr/88kvevHmTOTk5XLduHR0dHXnkyBGS5P79+5mens579+7xxx9/ZJ8+ffjqq68qGTIhIYEXL15kXl4eDx48SH9/fy5cuFAp7969O999912WlJTw5MmT1Gg03L9//5NI3kKIegaSLCkp4YABA9i0aVM6Ozuzbdu2/Oc//6lUOnnyJAGwpKRE+S46Opqenp50d3fnpEmTWF5eTvJBQurVqxe1Wi3VajVDQkK4bds2pV18fDxbtmxJR0dH+vr6csKECbx9+7ZSPm3aNHp7e9PR0ZEtW7bk3LlzqdfrlfJr166xX79+VKvVfOaZZxgbG1t/Z0cI8UTJk9pCCJshL9cKIWyGJCQhhM2QhCSEsBmSkIQQNkMSkhDCZkhCEkLYDElIQgibIQlJCGEzJCEJIWyGJCQhhM2QhCSEsBmSkIQQNkMSkhDCZkhCEkLYDElIQgibIQlJCGEzJCEJIWyGJCQhhM2QhCSEsBmSkIQQNkMSkhDCZkhCEkLYDElIQgibIQlJCGEzJCEJIWyGJCQhhM2QhCSEsBmSkIQQNuN/AQAseGSpCHBwAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log.mutations().standardize()\n",
    "log.mutations().render(\"%%inline\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATsAAABDCAYAAADj2qLfAAAACXBIWXMAAA3XAAAN1wFCKJt4AAAH9klEQVR4nO3dX0hUbR4H8O+ZGW0nc0uTRmJqvNBCoS21tRtJKV4MBiPUdEktQi03ChYy2Yv+aHTjTVC56YxaFllKr4v2ByJCpygI7K1oFkkjVlJeDX1dHU2POjPPXvR2tklHR6s1Pd8PnIszzznP/H5y+HLOmXGOJIQQICJa5DTzXQAR0f8Dw46IVIFhR0SqwLAjWiCOHz+OoKAgNDU1TTnucDhgNBqxdetWr3M8ePAAQUFBOHnypPJaamoqdDoddDodNBoNNBqNsp6ZmYmQkBBlXZIkaLVaZd1qtSrzxMfHw2g0wuFwTHrf3t5eHD58GCaTCUuWLIHBYEBycjJaWlp86uP8+fPKe2q1WkiSpKwbjcZp+1MIIvrhTUxMCJPJJCoqKkR6evqU21y4cEEUFBSIDRs2iNbW1im3SU1NFVVVVWLt2rXC6XROGt+3b584ePCg1zpMJpO4cePGpNftdrvYvHmzOHLkiLh48aLH2MDAgAgPDxdms1m0tbUJl8slenp6RElJidDr9cJut8+qj8uXL4v169fPur8Fe2Y3POLC+Ru/4sOoa75LIfqmpjq2GxsbERMTg7179+Lx48fo7e2dtJ/VakVWVhb279/vccb1SU9PD54+fYrs7Gxs2rQJd+7c+WY1WywWZGdnT/nepaWlEEKgoaEB69atg0ajgcFgQGFhIVpbWxEVFTWrPryZqb8FGXbDIy78KeMXnPjHv7Fh9y8Ykd3zXRLRN+Ht2C4vL0deXh78/f2RkZGB6upqj/2ePHkCvV6PjRs3Ijs7G3V1dRgbG/PYpqqqCpmZmfDz80NOTs6sgmQ6IyMjuHnzJrKyshAdHQ1JkjwuT5uampCWlgadTjdp37CwMGg0/4shX/rwZqb+NAAgSdKCWgLD/oLf/jMBxwcXfhtwIkCvnfeauHD5FsuXx7a1/le8ffsW7e3tSEpKAgDk5uaioqIC4rOvyFosFuTm5gIAQkJCEB8fj59//lkZd7vdqKysRE5ODgDAbDbj1atXePfunU9BMp3a2lps374dwcHBSn2fB01/fz9CQ0OV9Y6ODoSGhipLfn6+z31440t/OgAef7SF4MOoC2t+eoA/BgQiZLkWDs0SCJc832URfbUvj+2Daatx6sTf8f79eyVMAGB4eBjNzc3Ytm0b+vv7UV9fj8bGRhw7dgwA4HQ60dPTg8zMTADAvXv30NnZibi4OGUOWZZRWVmJ06dPf1XN5eXlaG1txYoVKwB8zBMhBM6ePYvAwECsWrUKnZ2dyvZr1qzBy5cvAQBlZWVoa2sDAJ/68MaX/hbkZWyAXouuBz+h+K9r8a9//plBR4vGl8e2VnKiuroadrsdAwMDynLu3Dnl7OnKlSvYtWsXBgcHlXGHw4E3b97g9evXAD6eMZWVlXnM8fz5c1y6dAku19zve7948QLd3d0YGhpS5h0cHERSUhJqamoAAElJSairq8Po6CgAQKvVKmd1y5YtU+bypQ9vfOlvQYYdACz9gwZ/yzRCv2TBtkA0pc+P7fr6ehiNRkRERHhsk5aWhlu3bqGvrw9WqxUZGRke4xqNBikpKbBarejq6sL9+/eRkpLisU1UVBSWL1+Ou3fvzrnW8vJy7N69G5Ikebyenp6uhHF+fj6Cg4OxY8cO2O12uN1u9PX1oaKiAiUlJYiJiQGAGfvwxtf+mBREPzCLxTIpAADAYDBgy5YtOHHiBLq7u5X7eZ9LT0/H1atXYbFYkJiYiJUrV065zVw/qBgaGsL169enrM9sNqO9vR3Pnj3D0qVL8ejRI8TGxmLnzp3Q6/WIjIxEY2MjampqUFhYCJvNNmMf3j6oqKys9Kk/SSy0G3ZERHPAMzsiUgWGHRGpgubhw4ew2WzzXcdXWyx9ENH3obHZbIsiJBZLH0T0ffAylohUgWFHRKrAsCMiVdB1dHSgo6MDxcXF813LV7HZbAgLC5vvMojoB6X8EMBi+G7xYuiBiL4PXVhYGEwmE4qKiua7lq9SXFzMsCMir3jPjohUgWFHRKrAsCMiVdAlJiYuintdi6UPIvo++BNPRKQKymVsaWkpYmJi4Ofnpzzwwpvm5mYkJCQgMDAQ4eHhk8anm6u7uxvJyckwGAyQJAldXV2zmjsvLw8RERGQJAnXrl2bVbNEpF5K2K1evRpFRUXYs2fPjDsFBATgwIEDOHPmzJTj082l0WhgNptRW1s7p7mjo6NhtVoRGRk5Y51ERJ8oD3L89PvtTU1NGB4ennanuLg4xMXFoaGhYcrx6eYyGAzIz8/HwMDAnOY+dOgQAMDf33/aGomIPsdPY4lIFRh2RKQKDDsiUgWGHRGpghJ2TqcTsizD5XLB5XJBlmU4nU4AHx9CW1BQoKy73W7IsoyJiQkIISDLMsbHx+HLXAAgy7LyDMixsTHIsqyMzTT3+Pg4ZFmGEAITExOQZRlut/s7/XmIaNEQvzt16pQA4LEcPXpUCCFES0uLACBGR0eFEEI0NzdP2jY2Nlb4MtfvX2KetAwNDfk0d0JCwqTx27dvCyKi6fA/KIhIFXjPjohUgWFHRKrAsCMiVWDYEZEqMOyISBUYdkSkCgw7IlIFhh0RqQLDjohUgWFHRKrAsCMiVWDYEZEqMOyISBUYdkSkCgw7IlIFhh0RqQLDjohUgWFHRKrAsCMiVWDYEZEqMOyISBUYdkSkCgw7IlIFhh0RqQLDjohUgWFHRKrAsCMiVWDYEZEq/BcA3o08ZlfZSgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log.mutations(prune_death=True).render(\"%%inline\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Refactoring logs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TreeLog(WeakLog):\n",
    "    \n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super().__init__(*args, **kwargs)\n",
    "        self.tree = Tree()\n",
    "        self.alive = {}\n",
    "        self.tmp = None\n",
    "        \n",
    "    def log_newcell(self, cell):\n",
    "        self.alive[cell.index] = self.tree\\\n",
    "                                    .add_child(name=cell.index)\n",
    "        \n",
    "    def preparefor_division(self, cell):\n",
    "        self.tmp = cell.index\n",
    "\n",
    "    def log_death(self, cell):\n",
    "        # No further mutations, so no need to hold it\n",
    "        del self.alive[cell.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "class AncestryLog(TreeLog):\n",
    "        \n",
    "    def log_division(self, daughters):\n",
    "        'Add 2 new branches to the father of the cells.'\n",
    "        d1, d2 = daughters\n",
    "        father = self.tmp\n",
    "        self.alive[d1.index] = self.alive[father]\\\n",
    "                                .add_child(name=d1.index)\n",
    "        self.alive[d2.index] = self.alive[father]\\\n",
    "                                .add_child(name=d2.index)\n",
    "        del self.alive[father]\n",
    "        self.tmp = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MutationsLog(TreeLog):\n",
    "\n",
    "    def log_division(self, daughters):\n",
    "        'Remove the father from the alive cells.'\n",
    "        d1, d2 = daughters\n",
    "        father = self.tmp\n",
    "        node = self.alive[father]\n",
    "        self.alive[d1.index] = node\n",
    "        self.alive[d2.index] = node\n",
    "        # Cleanup\n",
    "        del self.alive[father]\n",
    "        self.tmp = None\n",
    "        \n",
    "    def log_mutation(self, cell):\n",
    "        'Add a new child to the parent genome.'\n",
    "        node = self.alive[cell.index]\n",
    "        new = node.add_child(name=cell.genome)\n",
    "        self.alive[cell.index] = new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
